# **Ejercicios pares de la sección 10**
##### *Autor: Escorcia Alcantara Gregorio Jacobo*

****

### Ejercicios 10.1

Primero definiremos las constantes y variables globales, que son las de la tolerancia de error aceptada y el número máximo de iteraciones:

In [1]:
#Constantes y variables globales
TOL = 0.00001
const max_iter = 300

300

En seguida, continuaremos con la resolución de los ejercicios.

2.- El sistema no lineal

$$x_2^2 - x_1^2 + 4x_1 - 2 = 0, \hspace{0.2cm} x_1^2 + 3 x_2^2 - 4 =0$$

tiene dos soluciones.

**a.** Aproxime las soluciones gráficamente

**b.** Use las aproximaciones a partir de la parte **a)** como aproximaciones iniciales para una iteración de punto fijo adecuada y determine las soluciones dentro de $10^{-5}$ en la norma $l_\infty$.

**Resolución:**

In [2]:
# ----- FUNCIONES -----
#Funciones originales
f_1(x_1, x_2) = x_2^2 - x_1^2 + 4*x_1 - 2
f_2(x_1, x_2) = x_1^2 + 3*x_2^2 - 4

function F(x_1, x_2)
    
    f_1 = x_2^2 - x_1^2 + 4*x_1 - 2
    f_2 = x_1^2 + 3*x_2^2 - 4

    #println("$f_1, $f_2")

    return [f_1, f_2]
end

#Variables despejadas:
g_1(x_1, x_2) = (-x_2^2 + x_1^2 + 2)*(1/4) #x_1
g_2(x_1, x_2) = (-x_1^2 + 4)*(1/(3*x_2))   #x_2

#----------------------

g_2 (generic function with 1 method)

In [3]:
function puntoFijoMV(x_1n, x_2n)
    x_1m = 0.0
    x_2m = 0.0
    iteration = 1
    δ = 1.0

    #Ciclo para seguir calculando las raíces si el error es mayor al tolerado
    while (abs(δ) > TOL) && (iteration ≤ max_iter)

        x_1m = g_1(x_1n, x_2n)
        x_2m = g_2(x_1m, x_2n)

        #Comentar o descomentar la tabla en caso de querer o no los acercamientos a la raíz:
        println("X_$iteration = [$x_1m, $x_2m]")

        #Vectores:
        α = (x_1n, x_2n)
        β = (x_1m, x_2m)
        ω = abs.(β .- α)
        
        #Error
        δ = max(ω[1], ω[2])

        #Se asigna a x_n los valores de x_n+1
        x_1n = x_1m
        x_2n = x_2m

        iteration = iteration + 1
        
    end
    
    β = (x_1m, x_2m)
    println("Raíz encontrada X_$iteration = $β\n")
    resultado = F(x_1m, x_2m)
    println("Evaluando la función original con el resultado:\nF(X_$iteration) = $resultado")

end

puntoFijoMV (generic function with 1 method)

In [4]:
#Valores iniciales -> X_0
x_1n = 0.15
x_2n = 1.15

1.15

In [5]:
puntoFijoMV(x_1n, x_2n)

X_1 = [0.17500000000000004, 1.1505434782608697]
X_2 = [0.17671867615784498, 1.14982486812699]
X_3 = [0.1772830657849328, 1.150485557871814]
X_4 = [0.17695306663562138, 1.149858737195447]
X_5 = [0.1772843180717607, 1.150451541623086]
X_6 = [0.17697274495280846, 1.1498907179653286]
X_7 = [0.17726767229832818, 1.1504212560825258]
X_8 = [0.1769886902988928, 1.1499193541852721]
X_9 = [0.17725261884096105, 1.1503941543581246]
X_10 = [0.17700294512615855, 1.1499449825904902]
X_11 = [0.17723914489957276, 1.1503699004727461]
X_12 = [0.17701570164276398, 1.1499679189395202]
X_13 = [0.17722708600949727, 1.1503481951172878]
X_14 = [0.177027117501454, 1.1499884459662775]
X_15 = [0.17721629361873492, 1.150328770460364]
X_16 = [0.17703733364377722, 1.1500068167201847]
X_17 = [0.17720663475020143, 1.150311386825496]
X_18 = [0.17704647618477387, 1.1500232576800546]
X_19 = [0.17719799038110007, 1.150295829729789]
X_20 = [0.1770546579753392, 1.1500379715426463]
X_21 = [0.17719025398020993, 1.150281907234

4.- El sistema no lineal

$$5x_1^2 - x_2^2 = 0, \hspace{0.2cm} x_2-0.25(sen(x_1) + cos(x_2) = 0$$

tiene una solución cerca de $(\frac{1}{4}, \frac{1}{4})^t$.

**a.** Encuentre la función **G** y un conjunto $D$ en $\mathbb{R}^2$ de tal forma que **G**: $D \rightarrow \mathbb{R}^2$ y **G** tiene fijo único en $D$.

**b.** Aplique la iteración del punto fijo para aproximar la solución dentro de $10^{-5}$ en la norma $l_\infty$.

**c.** ¿El método de Gauss-Siedel acelera la convergencia?

**Resolución:**

In [6]:
# ----- FUNCIONES -----
#Funciones originales
f_1(x_1, x_2) = 5*x_1^2 - x_2^2
f_2(x_1, x_2) = x_2 - 0.25*(sin(x_1) + cos(x_2))

function F(x_1, x_2)
    
    f_1 = 5*x_1^2 - x_2^2
    f_2 = x_2 - 0.25*(sin(x_1) + cos(x_2))

    return f_1, f_2
end

#Variables despejadas:
g_1(x_1, x_2) = x_2/√5
g_2(x_1, x_2) = 0.25*(sin(x_1) + cos(x_2))

function G(x_1, x_2)
    g_1 = x_2/√5
    g_2 = 0.25*(sin(x_1) + cos(x_2))

    return g_1, g_2
end

#----------------------

G (generic function with 1 method)

In [7]:
function puntoFijoMV(x_1n, x_2n)
    x_1m = 0.0
    x_2m = 0.0
    iteration = 1
    δ = 1.0

    #Ciclo para seguir calculando las raíces si el error es mayor al tolerado
    while (abs(δ) > TOL) && (iteration ≤ max_iter)
        x_1m = g_1(x_1n, x_2n)
        x_2m = g_2(x_1n, x_2n)

        #Comentar o descomentar la tabla en caso de querer o no los acercamientos a la raíz:
        println("X_$iteration = [$x_1m, $x_2m]")

        #Vectores:
        α = (x_1n, x_2n)
        β = (x_1m, x_2m)
        ω = abs.(β .- α)
        
        #Error
        δ = max(ω[1], ω[2])

        #Se asigna a x_n los valores de x_n+1
        x_1n = x_1m
        x_2n = x_2m

        iteration = iteration + 1
        
    end
    
    if δ ≤ TOL
        β = (x_1m, x_2m)
        println("Raíz encontrada X_$iteration = $β\n")
        resultado = F(x_1m, x_2m)
        println("Evaluando la función original con el resultado:\nF(X_$iteration) = $resultado")

    else
        println("error")
    end
    
end

puntoFijoMV (generic function with 1 method)

In [8]:
#Valores iniciales -> X_0
x_1n = 0.25
x_2n = 0.25

0.25

In [9]:
puntoFijoMV(x_1n, x_2n)

X_1 = [0.11180339887498948, 0.3040790952412919]
X_2 = [0.1359883054992323, 0.2664234275354251]
X_3 = [0.11914817895353995, 0.27507206819663893]
X_4 = [0.12301596863982853, 0.270318023474851]
X_5 = [0.12088989520663015, 0.2715979896943834]
X_6 = [0.12146231350178573, 0.2709847718373571]
X_7 = [0.1211880741391202, 0.271167908932786]
X_8 = [0.12126997553803637, 0.2710875934751832]
X_9 = [0.12123405737346761, 0.2711132960358651]
X_10 = [0.12124555190804372, 0.2711026616886501]
X_11 = [0.1212407960833899, 0.2711062261954329]
Raíz encontrada X_12 = (0.1212407960833899, 0.2711062261954329)

Evaluando la función original con el resultado:
F(X_12) = (-1.932707258567312e-6, 1.4188679919313607e-6)


6.- Use la iteración de punto fijo para encontrar soluciones para los siguientes sistemas no lineales, precisos dentro de $10^{-5}$ por medio de la norma $l_\infty$

**a.** $$x_2^2 + x_2^2 - x_1 = 0$$

$$x_1^2-x_2^2-x_2 =0$$

**Resolución:**

In [10]:
# ----- FUNCIONES -----
#Funciones originales
f_1(x_1, x_2) = 2*x_2^2 - x_1
f_2(x_1, x_2) = x_1^2 - x_2^2 - x_2

function F(x_1, x_2)
    
    f_1 = 2*x_2^2 - x_1
    f_2 = x_1^2 - x_2^2 - x_2

    return f_1, f_2
end

#Variables despejadas:
g_1(x_1, x_2) = 2*x_2^2
g_2(x_1, x_2) = x_1^2 - x_2^2

#----------------------

g_2 (generic function with 1 method)

In [11]:
function puntoFijoMV(x_1n, x_2n)
    x_1m = 0.0
    x_2m = 0.0
    iteration = 0
    δ = 1.0

    #Ciclo para seguir calculando las raíces si el error es mayor al tolerado
    while (abs(δ) > TOL) && (iteration ≤ max_iter)
        x_1m = g_1(x_1n, x_2n)
        x_2m = g_2(x_1n, x_2n)

        #Comentar o descomentar la tabla en caso de querer o no los acercamientos a la raíz:
        println("X_$iteration = [$x_1m, $x_2m]")

        #Vectores:
        α = (x_1n, x_2n)
        β = (x_1m, x_2m)
        ω = abs.(β .- α)
        
        #Error
        δ = max(ω[1], ω[2])

        #Se asigna a x_n los valores de x_n+1
        x_1n = x_1m
        x_2n = x_2m

        iteration = iteration + 1
        
    end
    
    if δ ≤ TOL
        β = (x_1m, x_2m)
        println("Raíz encontrada X_$iteration = $β\n")
        resultado = F(x_1m, x_2m)
        println("Evaluando la función original con el resultado:\nF(X_$iteration) = $resultado")

    else
        println("error")
    end
    
end

puntoFijoMV (generic function with 1 method)

In [12]:
#Valores iniciales -> X_0
x_1n = 0.25
x_2n = 0.25

0.25

In [13]:
puntoFijoMV(x_1n, x_2n)

X_0 = [0.125, 0.0]
X_1 = [0.0, 0.015625]
X_2 = [0.00048828125, -0.000244140625]
X_3 = [1.1920928955078125e-7, 1.7881393432617188e-7]
X_4 = [6.394884621840902e-14, -1.7763568394002505e-14]
Raíz encontrada X_5 = (6.394884621840902e-14, -1.7763568394002505e-14)

Evaluando la función original con el resultado:
F(X_5) = (-6.394884621840839e-14, 1.776356839400628e-14)


**b.** $$3x_1^2 - x_2^2 = 0$$
    $$ 3x_1x_2^2 - x_1^3 - 1=0$$

In [14]:
# ----- FUNCIONES -----
#Funciones originales
f_1(x_1, x_2) = 3*x_1^2 - x_2^2
f_2(x_1, x_2) = 3*x_1*x_2^2 - x_1^3 - 1

function F(x_1, x_2)
    
    f_1 = 3*x_1^2 - x_2^2
    f_2 = 3*x_1*x_2^2 - x_1^3 - 1

    return f_1, f_2
end

#Variables despejadas:
g_1(x_1, x_2) = abs(x_2)/√3  #-> x_1n
g_2(x_1, x_2) = √((x_1^3 + 1)/(3*x_1)) #-> x_2n

#----------------------

g_2 (generic function with 1 method)

In [15]:
function puntoFijoMV(x_1n, x_2n)
    x_1m = 0.0
    x_2m = 0.0
    iteration = 0
    δ = 1.0

    #Ciclo para seguir calculando las raíces si el error es mayor al tolerado
    while (abs(δ) > TOL) && (iteration ≤ max_iter)
        x_1m = g_1(x_1n, x_2n)
        x_2m = g_2(x_1n, x_2n)

        #Comentar o descomentar la tabla en caso de querer o no los acercamientos a la raíz:
        println("X_$iteration = [$x_1m, $x_2m]")

        #Vectores:
        α = (x_1n, x_2n)
        β = (x_1m, x_2m)
        ω = abs.(β .- α)
        
        #Error
        δ = max(ω[1], ω[2])

        #Se asigna a x_n los valores de x_n+1
        x_1n = x_1m
        x_2n = x_2m

        iteration = iteration + 1
        
    end
    
    if δ ≤ TOL
        β = (x_1m, x_2m)
        println("Raíz encontrada X_$iteration = $β\n")
        resultado = F(x_1m, x_2m)
        println("Evaluando la función original con el resultado:\nF(X_$iteration) = $resultado")

    else
        println("error -> δ = $δ")
    end
    
end

puntoFijoMV (generic function with 1 method)

In [16]:
#Valores iniciales -> X_0
x_1n = 0.6
x_2n = 0.8

0.8

In [17]:
puntoFijoMV(x_1n, x_2n)

X_0 = [0.46188021535170065, 0.8219218670625302]
X_1 = [0.4745368112013917, 0.8903925806059613]
X_2 = [0.5140683960972974, 0.8817602208544748]
X_3 = [0.509084500871035, 0.8582021436729556]
X_4 = [0.4954832386686948, 0.860906009042669]
X_5 = [0.49704431606775135, 0.8686647573494648]
X_6 = [0.5015238314912545, 0.8677453720222134]
X_7 = [0.5009930241250768, 0.8651491870664215]
X_8 = [0.4994941160419844, 0.8654535972272098]
X_9 = [0.4996698673302596, 0.866317870302196]
X_10 = [0.5001688556227561, 0.8662161738414386]
X_11 = [0.5001101411437623, 0.8659279588327424]
X_12 = [0.49994374013090703, 0.8659618324397028]
X_13 = [0.49996329706700404, 0.8660578903084721]
X_14 = [0.5000187561033957, 0.8660465963067897]
X_15 = [0.5000122355084842, 0.866014575484699]
X_16 = [0.49999374827823045, 0.866018339840808]
X_17 = [0.4999959216302433, 0.8660290132778594]
X_18 = [0.5000020839419982, 0.866027758457924]
Raíz encontrada X_19 = (0.5000020839419982, 0.866027758457924)

Evaluando la función original con e

**c.** $$x_1^2+x_2-37=0$$

$$x_1-x_2^2-5=0$$

$$x_1+x_2+x_3-3=0$$

**Resolución:**

In [18]:
# ----- FUNCIONES -----
#Funciones originales
f_1(x_1, x_2, x_3) = x_1^2 + x_2 - 37
f_2(x_1, x_2, x_3) = x_1 - x_2^2 - 5
f_3(x_1, x_2, x_3) = x_1 + x_2 + x_3 - 3

function F(x_1, x_2, x_3)
    
    f_1 = x_1^2 + x_2 - 37
    f_2 = x_1 - x_2^2 - 5
    f_3 = x_1 + x_2 + x_3 - 3

    return f_1, f_2, f_3
end

#Variables despejadas:
g_1(x_1, x_2, x_3) = √(-x_2 + 37)
g_2(x_1, x_2, x_3) = √(x_1 - 5)
g_3(x_1, x_2, x_3) = -x_1 - x_2 + 3

#----------------------

g_3 (generic function with 1 method)

In [19]:
function puntoFijoMV(x_1n, x_2n, x_3n)
    x_1m = 0.0
    x_2m = 0.0
    x_3m = 0.0
    iteration = 0
    δ = 1.0

    #Ciclo para seguir calculando las raíces si el error es mayor al tolerado
    while (abs(δ) > TOL) && (iteration ≤ max_iter)
        x_1m = g_1(x_1n, x_2n, x_3n)
        x_2m = g_2(x_1n, x_2n, x_3n)
        x_3m = g_3(x_1n, x_2n, x_3n)

        #Comentar o descomentar la tabla en caso de querer o no los acercamientos a la raíz:
        println("X_$iteration = [$x_1m, $x_2m, $x_3m]")

        #Vectores:
        α = (x_1n, x_2n, x_3n)
        β = (x_1m, x_2m, x_3m)
        ω = abs.(β .- α)
        
        #Error
        δ = max(ω[1], ω[2], ω[3])

        #Se asigna a x_n los valores de x_n+1
        x_1n = x_1m
        x_2n = x_2m
        x_3n = x_3m

        iteration = iteration + 1
        
    end
    
    β = (x_1m, x_2m, x_3m)
    println("Raíz encontrada X_$iteration = $β\n")
    resultado = F(x_1m, x_2m, x_3m)
    println("Evaluando la función original con el resultado:\nF(X_$iteration) = $resultado")

end

puntoFijoMV (generic function with 2 methods)

In [20]:
#Valores iniciales -> X_0
x_1n = 5.5
x_2n = 1
x_3n = -3

-3

In [21]:
puntoFijoMV(x_1n, x_2n, x_3n)

X_0 = [6.0, 0.7071067811865476, -3.5]
X_1 = [6.024358324237815, 1.0, -3.707106781186548]
X_2 = [6.0, 1.0121058858824084, -4.024358324237815]
X_3 = [5.998991091351744, 1.0, -4.012105885882408]
X_4 = [6.0, 0.9994954183745637, -3.998991091351744]
X_5 = [6.000042048321448, 1.0, -3.999495418374564]
X_6 = [6.0, 1.000021023939721, -4.000042048321448]
X_7 = [5.9999982480047676, 1.0, -4.000021023939721]
X_8 = [6.0, 0.9999991240020001, -3.9999982480047676]
X_9 = [6.000000072999833, 1.0, -3.9999991240020005]
Raíz encontrada X_10 = (6.000000072999833, 1.0, -3.9999991240020005)

Evaluando la función original con el resultado:
F(X_10) = (8.759979976957766e-7, 7.299983284525524e-8, 9.489978323173887e-7)


**d.** $$x_1^2+2x_2^2-x_2-2x_3=0$$

$$x_1^2-8x_2^2-10x_3=0$$

$$\frac{x_1^2}{7x_2x_3}-1=0$$

**Resolución:**


In [22]:
# ----- FUNCIONES -----
#Funciones originales
f_1(x_1, x_2, x_3) = x_1^2 + 2*x_2^2 - x_2 - 2*x_3
f_2(x_1, x_2, x_3) = x_1^2 - 8*x_2^2 + 10*x_3
f_3(x_1, x_2, x_3) = (x_1^2)/(7*x_2*x_3) - 1

function F(x_1, x_2, x_3)
    
    f_1 = x_1^2 + 2*x_2^2 - x_2 - 2*x_3
    f_2 = x_1^2 - 8*x_2^2 + 10*x_3
    f_3 = (x_1^2)/(7*x_2*x_3) - 1

    return f_1, f_2, f_3
end

#Variables despejadas:
g_1(x_1, x_2, x_3) = x_1^2 + 2*x_2^2 - 2*x_3 #-> x_2
g_2(x_1, x_2, x_3) = √(8*x_2^2 - 10*x_3) #-> x_1
g_3(x_1, x_2, x_3) = (x_1^2)/(7*x_2) #-> x_3

#----------------------

g_3 (generic function with 1 method)

In [23]:
function puntoFijoMV(x_1n, x_2n, x_3n)
    x_1m = 0.0
    x_2m = 0.0
    x_3m = 0.0
    iteration = 0
    δ = 1.0

    #Ciclo para seguir calculando las raíces si el error es mayor al tolerado
    while (abs(δ) > TOL) && (iteration ≤ max_iter)
        x_1m = g_2(x_1n, x_2n, x_3n)
        x_2m = g_1(x_1n, x_2n, x_3n)
        x_3m = g_3(x_1n, x_2n, x_3n)

        #Comentar o descomentar la tabla en caso de querer o no los acercamientos a la raíz:
        println("X_$iteration = [$x_1m, $x_2m, $x_3m]")

        #Vectores:
        α = (x_1n, x_2n, x_3n)
        β = (x_1m, x_2m, x_3m)
        ω = abs.(β .- α)
        
        #Error
        δ = max(ω[1], ω[2], ω[3])

        #Se asigna a x_n los valores de x_n+1
        x_1n = x_1m
        x_2n = x_2m
        x_3n = x_3m

        iteration = iteration + 1
        
    end
    
    β = (x_1m, x_2m, x_3m)
    println("Raíz encontrada X_$iteration = $β\n")
    resultado = F(x_1m, x_2m, x_3m)
    println("Evaluando la función original con el resultado:\nF(X_$iteration) = $resultado")

end

puntoFijoMV (generic function with 2 methods)

In [24]:
#Valores iniciales
x_1n = 1.0
x_2n = -1.0
x_3n = 1.0

1.0

In [25]:
##puntoFijoMV(x_1n, x_2n, x_3n)

No se pudo encontrar el resultado.

****

### Ejercicios 10.2

Para estos ejerciicos nos auxiliaremos del método de Gauss-Jordan, para resolución de sistemas lineales, el cuál lo definiremos en seguida, sin embargo, no lo ocuparemos en todo momento, pues Julia nos permite obtener la inversa de una matriz de forma más directa, facilitando así los cálculos para obtener las raíces de un sistema.

In [26]:
function metodoGaussJordan(sistema, vectorResul)
    auxD = 0.0
    auxEscal = 0.0

    for i in 1:convert(Int64, size(sistema)[1])

        #Convirtiendo el valor x_ii en uno para escalonar
        auxD = sistema[i,i]
        if auxD != 0
            for dividir in 1:convert(Int64, size(sistema)[1])
                sistema[i, dividir] = sistema[i, dividir] ./ auxD
            end
            vectorResul[i] = vectorResul[i] ./ auxD
        end

        #Proceso para escalonar la matriz
        for j in i+1:convert(Int64, size(sistema)[1])
            auxEscal = sistema[j, i] #Obteniendo los valores

            for k in 1:convert(Int64, size(sistema)[1]) #Convirtiendo en cero los valores debajo de la diagonal principal
                sistema[j, k] = sistema[j,k] - (auxEscal*sistema[i, k])
            end
            vectorResul[j] = vectorResul[j] - (auxEscal*vectorResul[i])
        end
    end

    #Proceso inverso para convertir el sistema dado en una matriz diagonal
    for i in convert(Int64, size(sistema)[1]):-1:1

        for j in i-1:-1:1
            pivote = sistema[j, i] #Obteniendo los valores

            for k in convert(Int64, size(sistema)[1]):-1:1 #Convirtiendo en cero los valores arriba de la diagonal principal
                sistema[j, k] = sistema[j,k] - (pivote*sistema[i, k])
            end
            vectorResul[j] = vectorResul[j] - (pivote*vectorResul[i])
        end
    end

    #Regresa el valor del vector resultado obtenido de la resolución del sistema
    return vectorResul
end

metodoGaussJordan (generic function with 1 method)

2.- Use el método de Newton con $x^{0} = 0$ para calcular $x^{2}$ para cada uno de los siguientes sistemas no lineales.

**a.** $$3x_1-cos(x_2x_3)-\frac{1}{2}=0$$

$$4x_1^2-625x_2^2+2x_2-1=0$$

$$e^{-x_1x_2}+20x_3+\frac{10\pi-3}{3}=0$$

**Resolución:**

In [27]:
# ----- FUNCIONES -----
#Funciones originales
f_1(x_1, x_2, x_3) = 3*x_1 - cos(x_2*x_3) - 1/2
f_2(x_1, x_2, x_3) = 4*x_1^2 - 625*x_2^2 + 2*x_2 -1
f_3(x_1, x_2, x_3) = ℯ^(-x_1*x_2) + 20*x_3 + (10π-3)/3

function F(x_1, x_2, x_3)

    f_1 = 3*x_1 - cos(x_2*x_3) - 1/2
    f_2 = 4*x_1^2 - 625*x_2^2 + 2*x_2 -1
    f_3 = ℯ^(-x_1*x_2) + 20*x_3 + (10π-3)/3

    return [f_1; f_2; f_3]
end
#----------------------

F (generic function with 2 methods)

In [28]:
#Matriz Jacobiana:
function J(x_1, x_2, x_3)
    a_11 = 3
    a_12 = x_3*sin(x_2*x_3)
    a_13 = x_2*sin(x_2*x_3)

    a_21 = 8*x_1
    a_22 = -1250*x_2 + 2
    a_23 = 0

    a_31 = -x_2*ℯ^(-x_1*x_2)
    a_32 = -x_1*ℯ^(-x_1*x_2)
    a_33 = 20

    return [a_11 a_12 a_13; a_21 a_22 a_23; a_31 a_32 a_33]
end

J (generic function with 1 method)

In [29]:
function metodoNewton(x_1n, x_2n, x_3n)
    vector_xm = 0.0
    iteration = 0
    δ = 1.0

    while iteration ≤ 1 && δ ≥ TOL
        vectorFx = F(x_1n, x_2n, x_3n)  #Vector resultante al evaluar X_n en la función
        matrizJ = J(x_1n, x_2n, x_3n)   #Matriz Jacobiana evaluada en el vector X_n

        #Se obtiene el resultado equivalente de multiplicar
        #la matriz Jacobiana invertida multiplicada por el 
        #vector resultante al evaluar F(X_n), recordando que 
        #esto es lo mismo que resolver un sistema de ecuaciones 
        #J(X_n)*Y = F(X_n), en donde se busca al vector Y:
        #vectorResul = metodoGaussJordan(matrizJ, vectorFx)
        #Sin embargo, obtenemos el mismo resultado haciendo esto:
        vectorResul = inv(matrizJ)*vectorFx

        vector_xn = [x_1n, x_2n, x_3n]

        vector_xm = vector_xn .- vectorResul    #Se obtiene a X_n+1 = X_n - Y

        #Error
        ω = abs.(vector_xm .- vector_xn)
        δ = max(ω[1], ω[2], ω[3])

        println("X_$iteration = $vector_xn")

        #Se iguala x_n a x_n+1 para la siguiente iteración
        x_1n = vector_xm[1]
        x_2n = vector_xm[2]
        x_3n = vector_xm[3]

        iteration = iteration + 1
    end

    println("\nRaíz encontrada X_$iteration = $vector_xm\n")
    println("Evaluando la función original con la raíz encontrada: ")
    resultadoF = F(vector_xm[1], vector_xm[2], vector_xm[3])
    println("F(X_$iteration) = $resultadoF")
end

metodoNewton (generic function with 1 method)

In [30]:
#Valores iniciales
x_1n = 0.0
x_2n = 0.0
x_3n = 0.0

0.0

In [31]:
metodoNewton(x_1n, x_2n, x_3n)

X_0 = [0.0, 0.0, 0.0]
X_1 = [0.5, 0.5, -0.5235987755982989]

Raíz encontrada X_2 = [0.5001666869114633, 0.2508036384392389, -0.5173874273924913]

Evaluando la función original con la raíz encontrada: 
F(X_2) = [0.008907426191865686, -38.81176652331286, 0.006332455923706348]


**b.** $$x_1^2+x_2-37=0$$

$$x_1-x_2^2-5=0$$

$$x_1+x_2+x_3-3=0$$

In [32]:
# ----- FUNCIONES -----
#Funciones originales
f_1(x_1, x_2, x_3) = x_1^2 + x_2 - 37
f_2(x_1, x_2, x_3) = x_1 - x_2^2 - 5
f_3(x_1, x_2, x_3) = x_1 + x_2 + x_3 - 3

function F(x_1, x_2, x_3)

    f_1 = x_1^2 + x_2 - 37
    f_2 = x_1 - x_2^2 - 5
    f_3 = x_1 + x_2 + x_3 - 3

    return [f_1; f_2; f_3]
end
#----------------------

F (generic function with 2 methods)

In [33]:
#Matriz Jacobiana:
function J(x_1, x_2, x_3)
    a_11 = 2*x_1
    a_12 = x_2
    a_13 = 0

    a_21 = 1
    a_22 = -2*x_2
    a_23 = 0

    a_31 = 1
    a_32 = 1
    a_33 = 1

    return [a_11 a_12 a_13; a_21 a_22 a_23; a_31 a_32 a_33]
end

J (generic function with 1 method)

In [34]:
function metodoNewton(x_1n, x_2n, x_3n)
    vector_xm = 0.0
    iteration = 0
    δ = 1.0

    while iteration ≤ 1 && δ ≥ TOL
        vectorFx = F(x_1n, x_2n, x_3n)  #Vector resultante al evaluar X_n en la función
        matrizJ = J(x_1n, x_2n, x_3n)   #Matriz Jacobiana evaluada en el vector X_n

        #Se obtiene el resultado equivalente de multiplicar
        #la matriz Jacobiana invertida multiplicada por el 
        #vector resultante al evaluar F(X_n), recordando que 
        #esto es lo mismo que resolver un sistema de ecuaciones 
        #J(X_n)*Y = F(X_n), en donde se busca al vector Y:
        vectorResul = metodoGaussJordan(matrizJ, vectorFx)
        #Sin embargo, obtenemos el mismo resultado haciendo esto:
        #vectorResul = inv(matrizJ)*vectorFx

        vector_xn = [x_1n, x_2n, x_3n]

        vector_xm = vector_xn .- vectorResul    #Se obtiene a X_n+1 = X_n - Y

        #Error
        ω = abs.(vector_xm .- vector_xn)
        δ = max(ω[1], ω[2], ω[3])

        println("X_$iteration = $vector_xn")

        #Se iguala x_n a x_n+1 para la siguiente iteración
        x_1n = vector_xm[1]
        x_2n = vector_xm[2]
        x_3n = vector_xm[3]

        iteration = iteration + 1
    end

    println("Raíz encontrada X_$iteration = $vector_xm\n")
    println("Evaluando la función original con la raíz encontrada: ")
    resultadoF = F(vector_xm[1], vector_xm[2], vector_xm[3])
    println("F(X_$iteration) = $resultadoF")
end

metodoNewton (generic function with 1 method)

In [35]:
#Valores iniciales
x_1n = 0.0
x_2n = 0.0
x_3n = 0.0

0.0

In [36]:
metodoNewton(x_1n, x_2n, x_3n)

X_0 = [0.0, 0.0, 0.0]
X_1 = [37.0, -32.0, -2.0]
Raíz encontrada X_2 = [26.20805369127517, -16.331375838926174, -6.876677852348994]

Evaluando la función original con la raíz encontrada: 
F(X_2) = [633.5307024458358, -245.50578310098643, 0.0]


**c.** $$15x_1+x_2^2-4x_3=13$$

$$x_1^2+10x_2-x_3=11$$

$$x_2^3-25x_3=-22$$

In [37]:
# ----- FUNCIONES -----
#Funciones originales
f_1(x_1, x_2, x_3) = 15*x_1 + x_2^2 - 4*x_3 - 13
f_2(x_1, x_2, x_3) = x_1^2 + 10*x_2 - x_3 - 11
f_3(x_1, x_2, x_3) = x_2^3 - 25*x_3 + 22

function F(x_1, x_2, x_3)

    f_1 = 15*x_1 + x_2^2 - 4*x_3 - 13
    f_2 = x_1^2 + 10*x_2 - x_3 - 11
    f_3 = x_2^3 - 25*x_3 + 22

    return [f_1; f_2; f_3]
end
#----------------------

F (generic function with 2 methods)

In [38]:
#Matriz Jacobiana:
function J(x_1, x_2, x_3)
    a_11 = 15
    a_12 = 2*x_2
    a_13 = -4

    a_21 = 2*x_1
    a_22 = 10
    a_23 = -1

    a_31 = 0
    a_32 = 3*x_2^2
    a_33 = -25

    return [a_11 a_12 a_13; a_21 a_22 a_23; a_31 a_32 a_33]
end

J (generic function with 1 method)

In [39]:
function metodoNewton(x_1n, x_2n, x_3n)
    vector_xm = 0.0
    iteration = 0
    δ = 1.0

    while iteration ≤ 1 && δ ≥ TOL
        vectorFx = F(x_1n, x_2n, x_3n)  #Vector resultante al evaluar X_n en la función
        matrizJ = J(x_1n, x_2n, x_3n)   #Matriz Jacobiana evaluada en el vector X_n

        #Se obtiene el resultado equivalente de multiplicar
        #la matriz Jacobiana invertida multiplicada por el 
        #vector resultante al evaluar F(X_n), recordando que 
        #esto es lo mismo que resolver un sistema de ecuaciones 
        #J(X_n)*Y = F(X_n), en donde se busca al vector Y:
        #vectorResul = metodoGaussJordan(matrizJ, vectorFx)
        #Sin embargo, obtenemos el mismo resultado haciendo esto:
        vectorResul = inv(matrizJ)*vectorFx

        vector_xn = [x_1n, x_2n, x_3n]

        vector_xm = vector_xn .- vectorResul    #Se obtiene a X_n+1 = X_n - Y

        #Error
        ω = abs.(vector_xm .- vector_xn)
        δ = max(ω[1], ω[2], ω[3])

        println("X_$iteration = $vector_xn")

        #Se iguala x_n a x_n+1 para la siguiente iteración
        x_1n = vector_xm[1]
        x_2n = vector_xm[2]
        x_3n = vector_xm[3]

        iteration = iteration + 1
    end

    println("\nRaíz encontrada X_$iteration = $vector_xm\n")
    println("Evaluando la función original con la raíz encontrada: ")
    resultadoF = F(vector_xm[1], vector_xm[2], vector_xm[3])
    println("F(X_$iteration) = $resultadoF")
end

metodoNewton (generic function with 1 method)

In [40]:
#Valores iniciales
x_1n = 0.0
x_2n = 0.0
x_3n = 0.0

0.0

In [41]:
metodoNewton(x_1n, x_2n, x_3n)

X_0 = [0.0, 0.0, 0.0]
X_1 = [1.1013333333333333, 1.1880000000000002, 0.88]

Raíz encontrada X_2 = [1.0366870824055885, 1.0859238347813478, 0.9297793168810775]

Evaluando la función original con la raíz encontrada: 
F(X_2) = [0.010419543505744144, 0.004179137759010132, 0.03607166601008416]


**d.** $$10x_1-2x_2^2+x_2-2x_3-5=0$$

$$8x_2^2+4x_3^2-9=0$$

$$8x_2x_3+4=0$$

In [42]:
# ----- FUNCIONES -----
#Funciones originales
f_1(x_1, x_2, x_3) = 10*x_1 - 2*x_2^2 + x_2 - 2*x_3 - 5
f_2(x_1, x_2, x_3) = 8*x_2^2 + 4*x_3^2 - 9
f_3(x_1, x_2, x_3) = 8*x_2*x_3 + 4

function F(x_1, x_2, x_3)

    f_1 = 10*x_1 - 2*x_2^2 + x_2 - 2*x_3 - 5
    f_2 = 8*x_2^2 + 4*x_3^2 - 9
    f_3 = 8*x_2*x_3 + 4

    return [f_1; f_2; f_3]
end
#----------------------

F (generic function with 2 methods)

In [43]:
#Matriz Jacobiana:
function J(x_1, x_2, x_3)
    a_11 = 10
    a_12 = -4*x_2 + 1
    a_13 = -2

    a_21 = 0
    a_22 = 16*x_2
    a_23 = 8*x_3

    a_31 = 0
    a_32 = 8*x_3
    a_33 = 8*x_2

    return [a_11 a_12 a_13; a_21 a_22 a_23; a_31 a_32 a_33]
end

J (generic function with 1 method)

In [44]:
function metodoNewton(x_1n, x_2n, x_3n)
    vector_xm = 0.0
    iteration = 0
    δ = 1.0

    while iteration ≤ 1 && δ ≥ TOL
        vectorFx = F(x_1n, x_2n, x_3n)  #Vector resultante al evaluar X_n en la función
        matrizJ = J(x_1n, x_2n, x_3n)   #Matriz Jacobiana evaluada en el vector X_n

        #Se obtiene el resultado equivalente de multiplicar
        #la matriz Jacobiana invertida multiplicada por el 
        #vector resultante al evaluar F(X_n), recordando que 
        #esto es lo mismo que resolver un sistema de ecuaciones 
        #J(X_n)*Y = F(X_n), en donde se busca al vector Y:
        vectorResul = metodoGaussJordan(matrizJ, vectorFx)
        #Sin embargo, obtenemos el mismo resultado haciendo esto:
        #vectorResul = inv(matrizJ)*vectorFx

        vector_xn = [x_1n, x_2n, x_3n]

        vector_xm = vector_xn .- vectorResul    #Se obtiene a X_n+1 = X_n - Y

        #Error
        ω = abs.(vector_xm .- vector_xn)
        δ = max(ω[1], ω[2], ω[3])

        println("X_$iteration = $vector_xn")

        #Se iguala x_n a x_n+1 para la siguiente iteración
        x_1n = vector_xm[1]
        x_2n = vector_xm[2]
        x_3n = vector_xm[3]

        iteration = iteration + 1
    end

    println("\nRaíz encontrada X_$iteration = $vector_xm\n")
    println("Evaluando la función original con la raíz encontrada: ")
    resultadoF = F(vector_xm[1], vector_xm[2], vector_xm[3])
    println("F(X_$iteration) = $resultadoF")
end

metodoNewton (generic function with 1 method)

In [45]:
#Valores iniciales
x_1n = 1.0
x_2n = -1.0
x_3n = 1.0

1.0

In [46]:
metodoNewton(x_1n, x_2n, x_3n)

X_0 = [1.0, -1.0, 1.0]
X_1 = [0.9375, -1.125, 0.375]

Raíz encontrada X_2 = [0.9006127450980392, -1.0134803921568627, 0.48161764705882354]

Evaluando la función original con la raíz encontrada: 
F(X_2) = [-0.02487324586697426, 0.14496227412533536, 0.09511966551326445]


4.- Use las facilidades de graficación de su CAS o calculadora para aproximar soluciones para los siguienes sistemas no lineales.

**a.** $$3x_1-cos(x_2x_3)-\frac{1}{2}=0$$

$$4x_1^2-625x_2^2+2x_2-1=0$$

$$e^{-x_1x_2}+20x_3+\frac{10\pi-3}{3}=0$$

$$-1 \leq x_1 \leq 1, -1 \leq x_2 \leq 1, -1 \leq x_3 \leq 1 $$

$X^2$ = [0.4992, 0.2510, -0.5190]

**b.** $$x_1^2+x_2-37=0$$

$$x_1-x_2^2-5=0$$

$$x_1+x_2+x_3-3=0$$

$$-4 \leq x_1 \leq 8, -2 \leq x_2 \leq 2, -6 \leq x_3 \leq 0 $$

$X^2$ = [3.1515, 30.4848, -30.6363]

**c.** $$15x_1 + x_2^2-4x_3=13$$

$$x_1^2+10x_2-x_3=11$$

$$x_2^3-25x_3=-22$$

$$0 \leq x_1 \leq 2, 0 \leq x_2 \leq 2, 0 \leq x_2 \leq 2 $$

$$y$$

$$0 \leq x_1 \leq 2, 0 \leq x_2 \leq 2, -2 \leq x_2 \leq 0 $$

$X^2$ = [1.0364, 1.0857, 0.9311] y $X^2$ = [1.0364, 1.0857, 0.9311]

**d.** $$10x_1-2x_2^2+x_2-2x_3-5=0$$

$$8x_2^2+4x_3^2-9=0$$

$$8x_2x_3+4=0$$

$$0 \leq x_1 \leq 2, -2 \leq x_2 \leq 0, 0 \leq x_2 \leq 2 $$

$X^2$ = [0.9006, -1.0134, 0.4816]

6.- Use las respuestas obtenidas en el ejercicio 4 como aproximaciones iniciales para el método de Newton. Itere  hasta $||x^k-x^{k-1}||_\infty < 10^{-6}$

In [47]:
#Redefinimos el valor para el error tolerado:
TOL = 0.000001

1.0e-6

**a.** $$3x_1-cos(x_2x_3)-\frac{1}{2}=0$$

$$4x_1^2-625x_2^2+2x_2-1=0$$

$$e^{-x_1x_2}+20x_3+\frac{10\pi-3}{3}=0$$

$$-1 \leq x_1 \leq 1, -1 \leq x_2 \leq 1, -1 \leq x_3 \leq 1 $$

In [48]:
# ----- FUNCIONES -----
#Funciones originales
f_1(x_1, x_2, x_3) = 3x_1 - cos(x_2*x_3) - 1/2
f_2(x_1, x_2, x_3) = 4*x_1^2 - 625(x_2)^2 + 2*x_2 - 1
f_3(x_1, x_2, x_3) = ℯ^(-x_1*x_2) + 20x_3 + (10π - 3)/3

function F(x_1, x_2, x_3)

    f_1 = 3x_1 - cos(x_2*x_3) - 1/2
    f_2 = 4*x_1^2 - 625(x_2)^2 + 2*x_2 - 1
    f_3 = ℯ^(-x_1*x_2) + 20x_3 + (10π - 3)/3

    return [f_1; f_2; f_3]
end
#----------------------

F (generic function with 2 methods)

In [49]:
#Matriz Jacobiana:
function J(x_1, x_2, x_3)
    a_11 = 3
    a_12 = x_3*(sin(x_2*x_3))
    a_13 = x_2*(sin(x_2*x_3))

    a_21 = 8*x_1
    a_22 = -1250*(x_2) + 2
    a_23 = 0

    a_31 = -1*x_2*(ℯ^(-1*x_1*x_2))
    a_32 = -1*x_1*(ℯ^(-1*x_1*x_2))
    a_33 = 20

    return [a_11 a_12 a_13; a_21 a_22 a_23; a_31 a_32 a_33]
end

J (generic function with 1 method)

In [50]:
function metodoNewton(x_1n, x_2n, x_3n)
    vector_xm = 0.0
    iteration = 0
    δ = 1.0

    while iteration ≤ max_iter && δ ≥ TOL
        vectorFx = F(x_1n, x_2n, x_3n)  #Vector resultante al evaluar X_n en la función
        matrizJ = J(x_1n, x_2n, x_3n)   #Matriz Jacobiana evaluada en el vector X_n
        #println("MJ = $matrizJ")

        #Se obtiene el resultado equivalente de multiplicar
        #la matriz Jacobiana invertida multiplicada por el 
        #vector resultante al evaluar F(X_n), recordando que 
        #esto es lo mismo que resolver un sistema de ecuaciones 
        #J(X_n)*Y = F(X_n), en donde se busca al vector Y:
        #vectorResul = metodoGaussJordan(matrizJ, vectorFx)
        #Sin embargo, obtenemos el mismo resultado haciendo esto:
        vectorResul = inv(matrizJ)*vectorFx

        vector_xn = [x_1n, x_2n, x_3n]

        vector_xm = vector_xn .- vectorResul    #Se obtiene a X_n+1 = X_n - Y

        #Error
        ω = abs.(vector_xm .- vector_xn)
        δ = max(ω[1], ω[2], ω[3])

        println("X_$iteration = $vector_xn")

        #Se iguala x_n a x_n+1 para la siguiente iteración
        x_1n = vector_xm[1]
        x_2n = vector_xm[2]
        x_3n = vector_xm[3]

        iteration = iteration + 1
    end

    println("Raíz encontrada X_$iteration = $vector_xm\n")
    println("Evaluando la función original con la raíz encontrada: ")
    resultadoF = F(vector_xm[1], vector_xm[2], vector_xm[3])
    println("F(X_$iteration) = $resultadoF")
end

metodoNewton (generic function with 1 method)

In [51]:
#Valores iniciales
x_1n = 0.4992
x_2n = 0.2510
x_3n = 0.5190

0.519

In [52]:
metodoNewton(x_1n, x_2n, x_3n)


X_0 = [0.4992, 0.251, 0.519]
X_1 = [0.511270377810799, 0.12644950062814822, -0.5203194454508077]
X_2 = [0.49998579824877476, 0.06403137478179302, -0.5220311693033741]
X_3 = [0.4999946794639086, 0.032835917189248776, -0.5227786540730212]
X_4 = [0.49999741098407124, 0.017258671826731416, -0.523167669515885]
X_5 = [0.4999986019375495, 0.009510794048256617, -0.5233611979651518]
X_6 = [0.4999991636032832, 0.005716862930577596, -0.5234559686376556]
X_7 = [0.4999994199191373, 0.003968896924950507, -0.5234996326064338]
X_8 = [0.4999995142399384, 0.003324128156878251, -0.5235157389017151]
X_9 = [0.4999995318610684, 0.0032035994167598984, -0.5235187497068174]
X_10 = [0.49999953252299284, 0.0031990711846066077, -0.5235188628219735]
Raíz encontrada X_11 = [0.49999953252392976, 0.0031990647749826117, -0.5235188629820858]

Evaluando la función original con la raíz encontrada: 
F(X_11) = [1.1102230246251565e-16, -2.5646151868841116e-14, -1.7763568394002505e-15]


**b.** $$x_1^2+x_2-37=0$$

$$x_1-x_2^2-5=0$$

$$x_1+x_2+x_3-3=0$$

$$-4 \leq x_1 \leq 8, -2 \leq x_2 \leq 2, -6 \leq x_3 \leq 0 $$

In [53]:
# ----- FUNCIONES -----
#Funciones originales
f_1(x_1, x_2, x_3) = x_1^2 + x_2 - 37
f_2(x_1, x_2, x_3) = x_1 - x_2^2 - 5
f_3(x_1, x_2, x_3) = x_1 + x_2 + x_3 - 3

function F(x_1, x_2, x_3)

    f_1 = x_1^2 + x_2 - 37
    f_2 = x_1 - x_2^2 - 5
    f_3 = x_1 + x_2 + x_3 - 3

    return [f_1; f_2; f_3]
end
#----------------------

F (generic function with 2 methods)

In [54]:
#Matriz Jacobiana:
function J(x_1, x_2, x_3)
    a_11 = 2*x_1
    a_12 = 1
    a_13 = 0

    a_21 = 1
    a_22 = -2*x_2
    a_23 = 0

    a_31 = 1
    a_32 = 1
    a_33 = 1

    return [a_11 a_12 a_13; a_21 a_22 a_23; a_31 a_32 a_33]
end

J (generic function with 1 method)

In [55]:
function metodoNewton(x_1n, x_2n, x_3n)
    vector_xm = 0.0
    iteration = 0
    δ = 1.0

    while iteration ≤ max_iter && δ ≥ TOL
        vectorFx = F(x_1n, x_2n, x_3n)  #Vector resultante al evaluar X_n en la función
        matrizJ = J(x_1n, x_2n, x_3n)   #Matriz Jacobiana evaluada en el vector X_n
        #println("MJ = $matrizJ")

        #Se obtiene el resultado equivalente de multiplicar
        #la matriz Jacobiana invertida multiplicada por el 
        #vector resultante al evaluar F(X_n), recordando que 
        #esto es lo mismo que resolver un sistema de ecuaciones 
        #J(X_n)*Y = F(X_n), en donde se busca al vector Y:
        #vectorResul = metodoGaussJordan(matrizJ, vectorFx)
        #Sin embargo, obtenemos el mismo resultado haciendo esto:
        vectorResul = inv(matrizJ)*vectorFx

        vector_xn = [x_1n, x_2n, x_3n]

        vector_xm = vector_xn .- vectorResul    #Se obtiene a X_n+1 = X_n - Y

        #Error
        ω = abs.(vector_xm .- vector_xn)
        δ = max(ω[1], ω[2], ω[3])

        println("X_$iteration = $vector_xn")

        #Se iguala x_n a x_n+1 para la siguiente iteración
        x_1n = vector_xm[1]
        x_2n = vector_xm[2]
        x_3n = vector_xm[3]

        iteration = iteration + 1
    end

    println("Raíz encontrada X_$iteration = $vector_xm\n")
    println("Evaluando la función original con la raíz encontrada: ")
    resultadoF = F(vector_xm[1], vector_xm[2], vector_xm[3])
    println("F(X_$iteration) = $resultadoF")
end

metodoNewton (generic function with 1 method)

In [56]:
#Valores iniciales
x_1n = 3.1515
x_2n = 30.4848
x_3n = -30.6363

-30.6363

In [57]:
metodoNewton(x_1n, x_2n, x_3n)

X_0 = [3.1515, 30.4848, -30.6363]
X_1 = [5.027621642141408, 15.242853039582698, -17.270474681724107]
X_2 = [5.434111485412367, 7.635666356791846, -10.069777842204214]
X_3 = [5.7655794166573004, 3.8679649796614615, -6.633544396318763]
X_4 = [5.913528335906538, 2.0520715031374763, -4.965599839044013]
X_5 = [5.978276790105014, 1.2643989832127858, -4.242675773317799]
X_6 = [5.997800162327959, 1.0267743748416631, -4.024574537169622]
X_7 = [5.999972421711485, 1.0003356574124442, -4.000308079123929]
X_8 = [5.999999995555642, 1.0000000540926135, -4.000000049648255]
Raíz encontrada X_9 = [6.0, 1.0000000000000013, -4.000000000000001]

Evaluando la función original con la raíz encontrada: 
F(X_9) = [0.0, -2.6645352591003757e-15, 8.881784197001252e-16]


**c.** $$15x_1 + x_2^2-4x_3=13$$

$$x_1^2+10x_2-x_3=11$$

$$x_2^3-25x_3=-22$$

$$0 \leq x_1 \leq 2, 0 \leq x_2 \leq 2, 0 \leq x_2 \leq 2 $$

$$y$$

$$0 \leq x_1 \leq 2, 0 \leq x_2 \leq 2, -2 \leq x_2 \leq 0 $$

In [58]:
# ----- FUNCIONES -----
#Funciones originales
f_1(x_1, x_2, x_3) = 15*x_1 + x_2^2 - 4*x_3 -13
f_2(x_1, x_2, x_3) = x_1^2 + 10*x_2 - x_3 - 11
f_3(x_1, x_2, x_3) = x_2^3 - 25*x_3 + 22

function F(x_1, x_2, x_3)

    f_1 = 15*x_1 + x_2^2 - 4*x_3 -13
    f_2 = x_1^2 + 10*x_2 - x_3 - 11
    f_3 = x_2^3 - 25*x_3 + 22

    return [f_1; f_2; f_3]
end
#----------------------

F (generic function with 2 methods)

In [59]:
#Matriz Jacobiana:
function J(x_1, x_2, x_3)
    a_11 = 15
    a_12 = 2*x_2
    a_13 = -4

    a_21 = 2*x_1
    a_22 = 10
    a_23 = -1

    a_31 = 0
    a_32 = 3*x_2^2
    a_33 = -25

    return [a_11 a_12 a_13; a_21 a_22 a_23; a_31 a_32 a_33]
end

J (generic function with 1 method)

In [60]:
function metodoNewton(x_1n, x_2n, x_3n)
    vector_xm = 0.0
    iteration = 0
    δ = 1.0

    while iteration ≤ max_iter && δ ≥ TOL
        vectorFx = F(x_1n, x_2n, x_3n)  #Vector resultante al evaluar X_n en la función
        matrizJ = J(x_1n, x_2n, x_3n)   #Matriz Jacobiana evaluada en el vector X_n
        #println("MJ = $matrizJ")

        #Se obtiene el resultado equivalente de multiplicar
        #la matriz Jacobiana invertida multiplicada por el 
        #vector resultante al evaluar F(X_n), recordando que 
        #esto es lo mismo que resolver un sistema de ecuaciones 
        #J(X_n)*Y = F(X_n), en donde se busca al vector Y:
        #vectorResul = metodoGaussJordan(matrizJ, vectorFx)
        #Sin embargo, obtenemos el mismo resultado haciendo esto:
        vectorResul = inv(matrizJ)*vectorFx

        vector_xn = [x_1n, x_2n, x_3n]

        vector_xm = vector_xn .- vectorResul    #Se obtiene a X_n+1 = X_n - Y

        #Error
        ω = abs.(vector_xm .- vector_xn)
        δ = max(ω[1], ω[2], ω[3])

        println("X_$iteration = $vector_xn")

        #Se iguala x_n a x_n+1 para la siguiente iteración
        x_1n = vector_xm[1]
        x_2n = vector_xm[2]
        x_3n = vector_xm[3]

        iteration = iteration + 1
    end

    println("Raíz encontrada X_$iteration = $vector_xm\n")
    println("Evaluando la función original con la raíz encontrada: ")
    resultadoF = F(vector_xm[1], vector_xm[2], vector_xm[3])
    println("F(X_$iteration) = $resultadoF")
end

metodoNewton (generic function with 1 method)

In [61]:
#Valores iniciales primer rango -> X_0
x_1n = 1.0364
x_2n = 1.0857
x_3n = 0.9311

0.9311

In [62]:
metodoNewton(x_1n, x_2n, x_3n)

X_0 = [1.0364, 1.0857, 0.9311]
X_1 = [1.0364004703306722, 1.0857065507408261, 0.9311914423096784]
Raíz encontrada X_2 = [1.0364004703292111, 1.0857065507416779, 0.9311914423153898]

Evaluando la función original con la raíz encontrada: 
F(X_2) = [0.0, 0.0, 3.552713678800501e-15]


In [63]:
#Valores iniciales segundo rango -> X_0
x_1n = 1.0364
x_2n = 1.0857
x_3n = 0.9311


0.9311

In [64]:
metodoNewton(x_1n, x_2n, x_3n)

X_0 = [1.0364, 1.0857, 0.9311]
X_1 = [1.0364004703306722, 1.0857065507408261, 0.9311914423096784]
Raíz encontrada X_2 = [1.0364004703292111, 1.0857065507416779, 0.9311914423153898]

Evaluando la función original con la raíz encontrada: 
F(X_2) = [0.0, 0.0, 3.552713678800501e-15]


**d.** $$10x_1-2x_2^2+x_2-2x_3-5=0$$

$$8x_2^2+4x_3^2-9=0$$

$$8x_2x_3+4=0$$

$$0 \leq x_1 \leq 2, -2 \leq x_2 \leq 0, 0 \leq x_2 \leq 2 $$


In [65]:
# ----- FUNCIONES -----
#Funciones originales
f_1(x_1, x_2, x_3) = 10*x_1 - 2*x_2^2 + x_2 - 2*x_3 - 5
f_2(x_1, x_2, x_3) = 8x_2^2 + 4*x_3^2 - 9
f_3(x_1, x_2, x_3) = 8*x_2*x_3 + 4

function F(x_1, x_2, x_3)

    f_1 = 10*x_1 - 2*x_2^2 + x_2 - 2*x_3 - 5
    f_2 = 8x_2^2 + 4*x_3^2 - 9
    f_3 = 8*x_2*x_3 + 4

    return [f_1; f_2; f_3]
end
#----------------------

F (generic function with 2 methods)

In [66]:
#Matriz Jacobiana:
function J(x_1, x_2, x_3)
    a_11 = 10
    a_12 = -4*x_2 + 1
    a_13 = -2

    a_21 = 0
    a_22 = 16*x_2
    a_23 = 8*x_3

    a_31 = 0
    a_32 = 8*x_3
    a_33 = 8*x_2

    return [a_11 a_12 a_13; a_21 a_22 a_23; a_31 a_32 a_33]
end

J (generic function with 1 method)

In [67]:
function metodoNewton(x_1n, x_2n, x_3n)
    vector_xm = 0.0
    iteration = 0
    δ = 1.0

    while iteration ≤ max_iter && δ ≥ TOL
        vectorFx = F(x_1n, x_2n, x_3n)  #Vector resultante al evaluar X_n en la función
        matrizJ = J(x_1n, x_2n, x_3n)   #Matriz Jacobiana evaluada en el vector X_n
        #println("MJ = $matrizJ")

        #Se obtiene el resultado equivalente de multiplicar
        #la matriz Jacobiana invertida multiplicada por el 
        #vector resultante al evaluar F(X_n), recordando que 
        #esto es lo mismo que resolver un sistema de ecuaciones 
        #J(X_n)*Y = F(X_n), en donde se busca al vector Y:
        #vectorResul = metodoGaussJordan(matrizJ, vectorFx)
        #Sin embargo, obtenemos el mismo resultado haciendo esto:
        vectorResul = inv(matrizJ)*vectorFx

        vector_xn = [x_1n, x_2n, x_3n]

        vector_xm = vector_xn .- vectorResul    #Se obtiene a X_n+1 = X_n - Y

        #Error
        ω = abs.(vector_xm .- vector_xn)
        δ = max(ω[1], ω[2], ω[3])

        println("X_$iteration = $vector_xn")

        #Se iguala x_n a x_n+1 para la siguiente iteración
        x_1n = vector_xm[1]
        x_2n = vector_xm[2]
        x_3n = vector_xm[3]

        iteration = iteration + 1
    end

    println("Raíz encontrada X_$iteration = $vector_xm\n")
    println("Evaluando la función original con la raíz encontrada: ")
    resultadoF = F(vector_xm[1], vector_xm[2], vector_xm[3])
    println("F(X_$iteration) = $resultadoF")
end

metodoNewton (generic function with 1 method)

In [68]:
#Valores iniciales
x_1n = 0.9006
x_2n = -1.0134
x_3n = 0.4816


0.4816

In [69]:
metodoNewton(x_1n, x_2n, x_3n)

X_0 = [0.9006, -1.0134, 0.4816]
X_1 = [0.9000217790872787, -1.0002591945241275, 0.49963352271282824]
X_2 = [0.9000000091194721, -1.0000001038190733, 0.49999985317766016]
Raíz encontrada X_3 = [0.9000000000000015, -1.0000000000000167, 0.49999999999997646]

Evaluando la función original con la raíz encontrada: 
F(X_3) = [-2.220446049250313e-14, 1.723066134218243e-13, 1.2168044349891716e-13]


8.- El sistema no lineal:

$$4x_1 - x_2 + x_3 = x_1x_4$$

$$-x_1 + 3x_2 - 2x_3 = x_2x_4$$

$$x_1 - 2x_2 + 3x_3 = x_3x_4$$

$$x_1^2 + x_2^2 + x_3^2 = 1$$

tiene seis soluciones.

**a.** Muestre que si $(x_1, x_2, x_3, x_4)^t$ es una solución, entonces $(-x_1, -x_2, -x_3, x_4)^t$ es solución.

**b.** Use el método de Newton tres veces para aproximar todas las soluciones. Itere hasta $||x^k-x^{k-1}||_\infty < 10^{-5}$

In [70]:
#Redefinimos el error tolerado
TOL = 0.00001

1.0e-5

In [71]:
# ----- FUNCIONES -----
#Funciones originales
f_1(x_1, x_2, x_3, x_4) = 4*x_1 - x_2 + x_3 - x_1*x_4
f_2(x_1, x_2, x_3, x_4) = -x_1 + 3*x_2 - 2*x_3 - x_2*x_4
f_3(x_1, x_2, x_3, x_4) = x_1 - 2*x_2 + 3*x_3 - x_3*x_4
f_3(x_1, x_2, x_3, x_4) = x_1^2 + x_2^2 + x_3^2 - 1

function F(x_1, x_2, x_3, x_4)

    f_1 = 4*x_1 - x_2 + x_3 - x_1*x_4
    f_2 = -x_1 + 3*x_2 - 2*x_3 - x_2*x_4
    f_3 = x_1 - 2*x_2 + 3*x_3 - x_3*x_4
    f_4 = x_1^2 + x_2^2 + x_3^2 - 1

    return [f_1; f_2; f_3; f_4]
end
#----------------------

F (generic function with 3 methods)

In [72]:
#Matriz Jacobiana:
function J(x_1, x_2, x_3, x_4)
    a_11 = 4 - x_4
    a_12 = -1
    a_13 = 1
    a_14 = x_1

    a_21 = -1
    a_22 = 3 - x_4
    a_23 = -2
    a_24 = x_2

    a_31 = 1
    a_32 = -2
    a_33 = 3 - x_4
    a_34 = x_3

    a_41 = 2*x_1
    a_42 = 2*x_2
    a_43 = 2*x_3
    a_44 = 0

    return [a_11 a_12 a_13 a_14; a_21 a_22 a_23 a_24; a_31 a_32 a_33 a_34; a_41 a_42 a_43 a_44]
end

J (generic function with 2 methods)

In [73]:
function metodo_Newton(x_1n, x_2n, x_3n, x_4n)
    vector_xm = 0.0
    iteration = 0
    δ = 1.0

    while iteration ≤ max_iter && δ ≥ TOL
        vectorFx = F(x_1n, x_2n, x_3n, x_4n)  #Vector resultante al evaluar X_n en la función
        matrizJ = J(x_1n, x_2n, x_3n, x_4n)   #Matriz Jacobiana evaluada en el vector X_n

        #Se obtiene el resultado equivalente de multiplicar
        #la matriz Jacobiana invertida multiplicada por el 
        #vector resultante al evaluar F(X_n), recordando que 
        #esto es lo mismo que resolver un sistema de ecuaciones 
        #J(X_n)*Y = F(X_n), en donde se busca al vector Y:
        #vectorResul = metodoGaussJordan(matrizJ, vectorFx)
        #Sin embargo, obtenemos el mismo resultado haciendo esto:
        vectorResul = inv(matrizJ)*vectorFx

        vector_xn = [x_1n, x_2n, x_3n, x_4n]

        vector_xm = vector_xn .- vectorResul    #Se obtiene a X_1 = X_0 - Y

        #Error
        ω = abs.(vector_xm .- vector_xn)
        δ = max(ω[1], ω[2], ω[3])

        println("X_$iteration = $vector_xn")

        #Se iguala X_0 a X_1 para la siguiente iteración
        x_1n = vector_xm[1]
        x_2n = vector_xm[2]
        x_3n = vector_xm[3]
        x_4n = vector_xm[4]

        iteration = iteration + 1
    end

    println("Raíz encontrada X_$iteration = $vector_xm\n")
    println("Evaluando la función original con la raíz encontrada: ")
    resultadoF = F(vector_xm[1], vector_xm[2], vector_xm[3], vector_xm[4])
    println("F(X_$iteration) = $resultadoF")
end

metodo_Newton (generic function with 1 method)

Ahora, como tenemos seis soluciones, encontramos rangos en los cuáles nos dan diferentes raíces para el sistema dado, y a su vez, comprobamos el punto **a** al cambiar el signo del valor x_1, x_2 y x_3:

Primero tomamos un valor cercano a cero, para poder ver si existe una raíz cercana:

In [74]:
x_1n = 1.0
x_2n = 1.0
x_3n = 1.0
x_4n = 1.0 

1.0

In [75]:
metodo_Newton(x_1n, x_2n, x_3n, x_4n)

X_0 = [1.0, 1.0, 1.0, 1.0]
X_1 = [-2.220446049250313e-16, 1.0, 0.9999999999999998, 1.0]
X_2 = [2.220446049250315e-17, 0.75, 0.75, 1.0000000000000002]
X_3 = [-1.2860083368574706e-17, 0.7083333333333334, 0.7083333333333334, 1.0000000000000004]
X_4 = [-1.7976905367361844e-17, 0.7071078431372549, 0.7071078431372549, 1.000000000000001]
Raíz encontrada X_5 = [1.924121810066244e-17, 0.7071067811873449, 0.7071067811873449, 1.000000000000002]

Evaluando la función original con la raíz encontrada: 
F(X_5) = [9.178108436185318e-17, -1.5543122344752192e-15, -1.5543122344752192e-15, 2.255307052223543e-12]


Al ver que los valores tomados sí lograron alcanzar una raíz, ahora le cambiamos el signo a los tres primeros valores, y así observar si obtenemos o no una raíz igual a la anterior, pero con los tres primeros valores con el signo invertido:

In [76]:
x_1n = -1.0
x_2n = -1.0
x_3n = -1.0
x_4n = 1.0

1.0

In [77]:
metodo_Newton(x_1n, x_2n, x_3n, x_4n)

X_0 = [-1.0, -1.0, -1.0, 1.0]
X_1 = [2.220446049250313e-16, -1.0, -0.9999999999999998, 1.0]
X_2 = [-2.220446049250315e-17, -0.75, -0.75, 1.0000000000000002]
X_3 = [1.2860083368574706e-17, -0.7083333333333334, -0.7083333333333334, 1.0000000000000004]
X_4 = [1.7976905367361844e-17, -0.7071078431372549, -0.7071078431372549, 1.000000000000001]
Raíz encontrada X_5 = [-1.924121810066244e-17, -0.7071067811873449, -0.7071067811873449, 1.000000000000002]

Evaluando la función original con la raíz encontrada: 
F(X_5) = [-9.178108436185318e-17, 1.5543122344752192e-15, 1.5543122344752192e-15, 2.255307052223543e-12]


Y ahora, podemos ver que el punto **a** se cumple, pues (0.0000, 0.7071, 0.7071, 1.0) es solución, pero también lo es (-0.0000, -0.7071, -0.7071, 1.0)

Proseguimos buscamos otro rango de valores que nos puedan encontrar una raíz distinta a las ya obtenidas:

In [78]:
x_1n = -1.0
x_2n = -0.5
x_3n = 0.5
x_4n = 3

3

In [79]:
metodo_Newton(x_1n, x_2n, x_3n, x_4n)

X_0 = [-1.0, -0.5, 0.5, 3.0]
X_1 = [-0.8333333333333334, -0.4166666666666667, 0.4166666666666667, 3.0]
X_2 = [-0.8166666666666667, -0.4083333333333334, 0.4083333333333334, 2.9999999999999996]
X_3 = [-0.8164965986394558, -0.40824829931972795, 0.40824829931972795, 2.999999999999999]
Raíz encontrada X_4 = [-0.8164965809277261, -0.40824829046386313, 0.40824829046386313, 2.9999999999999987]

Evaluando la función original con la raíz encontrada: 
F(X_4) = [-8.881784197001252e-16, -4.440892098500626e-16, 4.440892098500626e-16, 4.440892098500626e-16]


A continuación, repetiremos el procedimiento anterior e invertiremos el signo de los tres primeros valores para poder encontrar otra solución al sistema:

In [80]:
x_1n = 1.0
x_2n = 0.5
x_3n = -0.5
x_4n = 3

3

In [81]:
metodo_Newton(x_1n, x_2n, x_3n, x_4n)

X_0 = [1.0, 0.5, -0.5, 3.0]
X_1 = [0.8333333333333334, 0.4166666666666667, -0.4166666666666667, 3.0]
X_2 = [0.8166666666666667, 0.4083333333333334, -0.4083333333333334, 2.9999999999999996]
X_3 = [0.8164965986394558, 0.40824829931972795, -0.40824829931972795, 2.999999999999999]
Raíz encontrada X_4 = [0.8164965809277261, 0.40824829046386313, -0.40824829046386313, 2.9999999999999987]

Evaluando la función original con la raíz encontrada: 
F(X_4) = [8.881784197001252e-16, 4.440892098500626e-16, -4.440892098500626e-16, 4.440892098500626e-16]


Nuevamente, podemos comprobar lo propuesto en el punto **a**.

Tomando nuevos valores para seguir encontrando más soluciones:

In [82]:
x_1n = 0.5
x_2n = -0.5
x_3n = 0.5
x_4n = 6

6

In [83]:
metodo_Newton(x_1n, x_2n, x_3n, x_4n)

X_0 = [0.5, -0.5, 0.5, 6.0]
X_1 = [0.5833333333333334, -0.5833333333333334, 0.5833333333333334, 6.0]
X_2 = [0.5773809523809524, -0.5773809523809523, 0.5773809523809523, 6.0]
X_3 = [0.5773502700049092, -0.5773502700049091, 0.5773502700049092, 6.0]
Raíz encontrada X_4 = [0.5773502691896257, -0.5773502691896257, 0.5773502691896258, 6.000000000000001]

Evaluando la función original con la raíz encontrada: 
F(X_4) = [-4.440892098500626e-16, 4.440892098500626e-16, -8.881784197001252e-16, 0.0]


Invirtiendo nuevamente los valores, tenemos que:

In [84]:
x_1n = -0.5
x_2n = 0.5
x_3n = -0.5
x_4n = 6

6

In [85]:
metodo_Newton(x_1n, x_2n, x_3n, x_4n)

X_0 = [-0.5, 0.5, -0.5, 6.0]
X_1 = [-0.5833333333333334, 0.5833333333333334, -0.5833333333333334, 6.0]
X_2 = [-0.5773809523809524, 0.5773809523809523, -0.5773809523809523, 6.0]
X_3 = [-0.5773502700049092, 0.5773502700049091, -0.5773502700049092, 6.0]
Raíz encontrada X_4 = [-0.5773502691896257, 0.5773502691896257, -0.5773502691896258, 6.000000000000001]

Evaluando la función original con la raíz encontrada: 
F(X_4) = [4.440892098500626e-16, -4.440892098500626e-16, 8.881784197001252e-16, 0.0]


De esa forma, hemos conseguido las seis soluciones del sistema:

$$S_1 = (0.0000, 0.7071, 0.7071, 1.0)^t$$

$$S_2 = (-0.0000, -0.7071, -0.7071, 1.0)^t$$

$$S_3 = (-0.8164, -0.4082, 0.4082, 2.9999)^t$$

$$S_4 = (0.8164, 0.4082, -0.4082, 2.9999)^t$$

$$S_5 = (0.5773, -0.5773, 0.5773, 6.0000)^t$$

$$S_6 = (-0.5773, 0.5773, -0.5773, 6.0000)^t$$



Y a su vez, también realizamos el punto **b**, pues aproximamos todas las soluciones con un error tolerado $< 10^{-5}$.

****

### Ejercicios 10.3

2.- Use el método de Broyden para calcular $x^2$ para cada uno de los siguientes sistemas no lineales.

Para utilizar el método de Broyden, necesitamos usar el método de Newton para la primer iteración, la cuál definiremos a continuación por razones prácticas.

In [86]:
function metodo_Newton(x_1n, x_2n, x_3n)
    vector_xm = 0.0
    iteration = 0

    while iteration ≤ 1
        vectorFx = F(x_1n, x_2n, x_3n)  #Vector resultante al evaluar X_n en la función
        matrizJ = J(x_1n, x_2n, x_3n)   #Matriz Jacobiana evaluada en el vector X_n

        #Se obtiene el resultado equivalente de multiplicar
        #la matriz Jacobiana invertida multiplicada por el 
        #vector resultante al evaluar F(X_n), recordando que 
        #esto es lo mismo que resolver un sistema de ecuaciones 
        #J(X_n)*Y = F(X_n), en donde se busca al vector Y:
        #vectorResul = metodoGaussJordan(matrizJ, vectorFx)
        #Sin embargo, obtenemos el mismo resultado haciendo esto:
        vectorResul = inv(matrizJ)*vectorFx

        vector_xn = [x_1n, x_2n, x_3n]

        vector_xm = vector_xn .- vectorResul    #Se obtiene a X_1 = X_0 - Y

        #Se iguala X_0 a X_1 para la siguiente iteración
        x_1n = vector_xm[1]
        x_2n = vector_xm[2]
        x_3n = vector_xm[3]

        iteration = iteration + 1
    end

    return vector_xm
end

metodo_Newton (generic function with 2 methods)

Además, utilizaremos la fórmula de Sherman-Morrison, la cuál definiremos a continuación:

In [87]:
function sherman_morrison(x_1n, x_2n, x_3n, x_1m, x_2m, x_3m, A_0)
    x0 = [x_1n, x_2n, x_3n]
    x1 = [x_1m, x_2m, x_3m]
    y_i = F(x_1m, x_2m, x_3m) .- F(x_1n, x_2n, x_3n)
    s_i = x1 .- x0

    #Producto externo:
    AxY = A_0 * y_i
    prodExtern = 0.0
    for j in 1:convert(Int64, size(s_i)[1])
        prodExtern = prodExtern + s_i[j] * AxY[j]
    end

    A_1 = A_0 + ((s_i - (A_0 * y_i)) * (transpose(s_i) * A_0)) / prodExtern
    return A_1
end

sherman_morrison (generic function with 1 method)

Ahora, continuaremos resolviendo los ejercicios.

**a.**     $$3x_1 - cos(x_2*x_3) - 1/2=0$$

$$4x_1^2 - 625(x_2)^2 + 2x_2 - 1=0$$

$$ℯ^{-x_1*x_2} + 20x_3 + \frac{10\pi-3}{3}=0$$

Use $x^0=(0,0,0)^t$

**Resolución:**

In [88]:
# ----- FUNCIONES -----
#Funciones originales
f_1(x_1, x_2, x_3) = 3*x_1 - cos(x_2*x_3) - 1/2
f_2(x_1, x_2, x_3) = 4*x_1^2 - 625(x_2)^2 + 2*x_2 - 1
f_3(x_1, x_2, x_3) = ℯ^(-x_1*x_2) + 20x_3 + (10π - 3)/3

function F(x_1, x_2, x_3)

    f_1 = 3x_1 - cos(x_2*x_3) - 1/2
    f_2 = 4*x_1^2 - 625(x_2)^2 + 2*x_2 - 1
    f_3 = ℯ^(-x_1*x_2) + 20x_3 + (10π - 3)/3

    return [f_1; f_2; f_3]
end
#----------------------

F (generic function with 3 methods)

In [89]:
#Matriz Jacobiana:
function J(x_1, x_2, x_3)
    a_11 = 3
    a_12 = x_3*(sin(x_2*x_3))
    a_13 = x_2*(sin(x_2*x_3))

    a_21 = 8*x_1
    a_22 = -1250*(x_2) + 2
    a_23 = 0

    a_31 = -1*x_2*(ℯ^(-1*x_1*x_2))
    a_32 = -1*x_1*(ℯ^(-1*x_1*x_2))
    a_33 = 20

    return [a_11 a_12 a_13; a_21 a_22 a_23; a_31 a_32 a_33]
end

J (generic function with 2 methods)

In [90]:
function metodo_Broyden(x_1n, x_2n, x_3n)
    iteration = 0
    δ = 1
    x_n = [x_1n, x_2n, x_3n]

    #Primer iteración con Newton -> x_1
    x_m = metodo_Newton(x_1n, x_2n, x_3n)

    mJacobi = inv(J(x_1n, x_2n, x_3n)) #-> J(x_0)^-1

    iteration = 1

    while iteration ≤ 1

        mJacobi = sherman_morrison(x_n[1], x_n[2], x_n[3], x_m[1], x_m[2], x_m[3], mJacobi) #-> ≈ J(x_iteration)^-1
        x_n = [x_m[1], x_m[2], x_m[3]] #-> x_n = x_iteration
        
        #Se consigue a x_iteration+1
        x_m = x_n - (mJacobi * F(x_n[1], x_n[2], x_n[3]))

        #Error
        ω = abs.(x_m .- x_n)
        δ = max(ω[1], ω[2], ω[3])

        iteration = iteration + 1
    end

    println("La raíz encontrada es X_$iteration = $x_m")
    println("Evaluando la función original con la raíz encontrada: ")
    resultadoF = F(x_m[1], x_m[2], x_m[3])
    println("F(X_$iteration) = $resultadoF")
end

metodo_Broyden (generic function with 1 method)

In [91]:
#Valores iniciales
x_1n = 0
x_2n = 0
x_3n = 0

0

In [92]:
metodo_Broyden(x_1n, x_2n, x_3n)

La raíz encontrada es X_2 = [0.500429324972278, -2.4347281668395393, -0.5245018240630789]
Evaluando la función original con la raíz encontrada: 
F(X_2) = [0.7117185471458221, -3709.80601729765, 2.363743294819759]


**b.** 
$$x_1^2 + x_2 - 37=0$$

$$x_1 - x_2^2 - 5=0$$

$$x_1 + x_2 + x_3 - 3=0$$

Use $x^0 = (0,0,0)^t$

In [93]:
# ----- FUNCIONES -----
#Funciones originales
f_1(x_1, x_2, x_3) = x_1^2 + x_2 - 37
f_2(x_1, x_2, x_3) = x_1 - x_2^2 - 5
f_3(x_1, x_2, x_3) = x_1 + x_2 + x_3 - 3

function F(x_1, x_2, x_3)

    f_1 = x_1^2 + x_2 - 37
    f_2 = x_1 - x_2^2 - 5
    f_3 = x_1 + x_2 + x_3 - 3

    return [f_1; f_2; f_3]
end
#----------------------

F (generic function with 3 methods)

In [94]:
#Matriz Jacobiana:
function J(x_1, x_2, x_3)
    a_11 = 2*x_1
    a_12 = 1
    a_13 = 0

    a_21 = 1
    a_22 = -2*x_2
    a_23 = 0

    a_31 = 1
    a_32 = 1
    a_33 = 1

    return [a_11 a_12 a_13; a_21 a_22 a_23; a_31 a_32 a_33]
end

J (generic function with 2 methods)

In [95]:
function metodo_Broyden(x_1n, x_2n, x_3n)
    iteration = 0
    δ = 1
    x_n = [x_1n, x_2n, x_3n]

    #Primer iteración con Newton -> x_1
    x_m = metodo_Newton(x_1n, x_2n, x_3n)

    mJacobi = inv(J(x_1n, x_2n, x_3n)) #-> J(x_0)^-1

    iteration = 1

    while iteration ≤ 1

        mJacobi = sherman_morrison(x_n[1], x_n[2], x_n[3], x_m[1], x_m[2], x_m[3], mJacobi) #-> ≈ J(x_iteration)^-1
        x_n = [x_m[1], x_m[2], x_m[3]] #-> x_n = x_iteration
        
        #Se consigue a x_iteration+1
        x_m = x_n - (mJacobi * F(x_n[1], x_n[2], x_n[3]))

        #Error
        ω = abs.(x_m .- x_n)
        δ = max(ω[1], ω[2], ω[3])

        iteration = iteration + 1
    end

    println("La raíz encontrada es X_$iteration = $x_m")
    println("Evaluando la función original con la raíz encontrada: ")
    resultadoF = F(x_m[1], x_m[2], x_m[3])
    println("F(X_$iteration) = $resultadoF")
end

metodo_Broyden (generic function with 1 method)

In [96]:
#Valores iniciales
x_1n = 0
x_2n = 0
x_3n = 0

0

In [97]:
metodo_Broyden(x_1n, x_2n, x_3n)

La raíz encontrada es X_2 = [-69.49549598603848, 41.1243181168741, 31.371177869164363]
Evaluando la función original con la raíz encontrada: 
F(X_2) = [4833.748280462364, -1765.705036563898, -1.4210854715202004e-14]


**c.** 
$$x_1^3 + x_1^2x_2 - x_1x_3 + 6=0$$

$$ℯ^{x_1} + ℯ^{x_2} - x_3=0$$

$$x_2^2 - 2x_1x_3 - 4=0$$

Use $x^0 = (-1,-2,1)^t$

In [98]:
# ----- FUNCIONES -----
#Funciones originales
f_1(x_1, x_2, x_3) = x_1^3 + x_1^2*x_2 - x_1*x_3 + 6
f_2(x_1, x_2, x_3) = ℯ^(x_1) + ℯ^(x_2) - x_3 
f_3(x_1, x_2, x_3) = x_2^2 - 2*x_1*x_3 - 4

function F(x_1, x_2, x_3)

    f_1 = x_1^3 + x_1^2*x_2 - x_1*x_3 + 6
    f_2 = ℯ^(x_1) + ℯ^(x_2) - x_3 
    f_3 = x_2^2 - 2*x_1*x_3 - 4

    return [f_1; f_2; f_3]
end
#----------------------

F (generic function with 3 methods)

In [99]:
#Matriz Jacobiana:
function J(x_1, x_2, x_3)
    a_11 = 3*x_1^2 + 2*x_2*x_1 - x_3
    a_12 = x_1^2
    a_13 = -x_1

    a_21 = ℯ^(x_1)
    a_22 = ℯ^(x_2)
    a_23 = -1

    a_31 = -2*x_3
    a_32 = 2*x_2
    a_33 = -2*x_1

    return [a_11 a_12 a_13; a_21 a_22 a_23; a_31 a_32 a_33]
end

J (generic function with 2 methods)

In [100]:
function metodo_Broyden(x_1n, x_2n, x_3n)
    iteration = 0
    δ = 1
    x_n = [x_1n, x_2n, x_3n]

    #Primer iteración con Newton -> x_1
    x_m = metodo_Newton(x_1n, x_2n, x_3n)

    mJacobi = inv(J(x_1n, x_2n, x_3n)) #-> J(x_0)^-1

    iteration = 1

    while iteration ≤ 1

        mJacobi = sherman_morrison(x_n[1], x_n[2], x_n[3], x_m[1], x_m[2], x_m[3], mJacobi) #-> ≈ J(x_iteration)^-1
        x_n = [x_m[1], x_m[2], x_m[3]] #-> x_n = x_iteration
        
        #Se consigue a x_iteration+1
        x_m = x_n - (mJacobi * F(x_n[1], x_n[2], x_n[3]))

        #Error
        ω = abs.(x_m .- x_n)
        δ = max(ω[1], ω[2], ω[3])

        iteration = iteration + 1
    end

    println("La raíz encontrada es X_$iteration = $x_m")
    println("Evaluando la función original con la raíz encontrada: ")
    resultadoF = F(x_m[1], x_m[2], x_m[3])
    println("F(X_$iteration) = $resultadoF")
end

metodo_Broyden (generic function with 1 method)

In [101]:
#Valores iniciales
x_1n = -1
x_2n = -2
x_3n = 1

1

In [102]:
metodo_Broyden(x_1n, x_2n, x_3n)

La raíz encontrada es X_2 = [-1.4503779511692627, -1.6693423055689351, 0.4240475133403323]
Evaluando la función original con la raíz encontrada: 
F(X_2) = [0.052397482597422496, -0.0011949492007941043, 0.01676206035615202]


**d.** 
$$6x_1 - 2cos(x_2x_3) - 1=0$$

$$9x_2 + √(x_1^2 + sin(x_3) + 1.06) + 0.9=0$$

$$60x_3 + 3ℯ^{-x_1*x_2} + 10π - 3=0$$

Use $x^0 = (0,0,0)^t$

In [103]:
# ----- FUNCIONES -----
#Funciones originales
f_1(x_1, x_2, x_3) = 6*x_1 - 2*cos(x_2*x_3) - 1
f_2(x_1, x_2, x_3) = 9*x_2 + √(x_1^2 + sin(x_3) + 1.06) + 0.9
f_3(x_1, x_2, x_3) = 60*x_3 + 3*ℯ^(-x_1*x_2) + 10π - 3

function F(x_1, x_2, x_3)

    f_1 = 6*x_1 - 2*cos(x_2*x_3) - 1
    f_2 = 9*x_2 + √(x_1^2 + sin(x_3) + 1.06) + 0.9
    f_3 = 60*x_3 + 3*ℯ^(-x_1*x_2) + 10π - 3

    return [f_1; f_2; f_3]
end
#----------------------

F (generic function with 3 methods)

In [104]:
#Matriz Jacobiana:
function J(x_1, x_2, x_3)
    a_11 = 6
    a_12 = 2*x_3*(sin(x_2*x_3))
    a_13 = 2*x_2*(sin(x_2*x_3))

    a_21 = x_1/(√(x_1^2 + sin(x_3) + 1.06))
    a_22 = 9
    a_23 = cos(x_3)/(2*√(x_1^2 + sin(x_3) + 1.06))

    a_31 = -3*x_2*ℯ^(-x_1*x_2)
    a_32 = -3*x_1*ℯ^(-x_1*x_2)
    a_33 = 60

    return [a_11 a_12 a_13; a_21 a_22 a_23; a_31 a_32 a_33]
end

J (generic function with 2 methods)

In [105]:
function metodo_Broyden(x_1n, x_2n, x_3n)
    iteration = 0
    δ = 1
    x_n = [x_1n, x_2n, x_3n]

    #Primer iteración con Newton -> x_1
    x_m = metodo_Newton(x_1n, x_2n, x_3n)

    mJacobi = inv(J(x_1n, x_2n, x_3n)) #-> J(x_0)^-1

    iteration = 1

    while iteration ≤ 1

        mJacobi = sherman_morrison(x_n[1], x_n[2], x_n[3], x_m[1], x_m[2], x_m[3], mJacobi) #-> ≈ J(x_iteration)^-1
        x_n = [x_m[1], x_m[2], x_m[3]] #-> x_n = x_iteration
        
        #Se consigue a x_iteration+1
        x_m = x_n - (mJacobi * F(x_n[1], x_n[2], x_n[3]))

        #Error
        ω = abs.(x_m .- x_n)
        δ = max(ω[1], ω[2], ω[3])

        iteration = iteration + 1
    end

    println("La raíz encontrada es X_$iteration = $x_m")
    println("Evaluando la función original con la raíz encontrada: ")
    resultadoF = F(x_m[1], x_m[2], x_m[3])
    println("F(X_$iteration) = $resultadoF")
end

metodo_Broyden (generic function with 1 method)

In [106]:
#Valores iniciales
x_1n = 0
x_2n = 0
x_3n = 0

0

In [107]:
metodo_Broyden(x_1n, x_2n, x_3n)

La raíz encontrada es X_2 = [0.49814468649719146, -0.19960653664721917, -0.5288260855863877]
Evaluando la función original con la raíz encontrada: 
F(X_2) = [8.743095736285511e-8, -5.82088610390219e-6, -5.421268973293536e-6]


4.- Use el método de Broyden para aproximar soluciones de los sistemas no lineales en el ejercicio 2 mediante las siguientes aproximaciones iniciales $x^0$.

**a.** $(1,1,1)^t$

In [108]:
# ----- FUNCIONES -----
#Funciones originales
f_1(x_1, x_2, x_3) = 3*x_1 - cos(x_2*x_3) - 1/2
f_2(x_1, x_2, x_3) = 4*x_1^2 - 625(x_2)^2 + 2*x_2 - 1
f_3(x_1, x_2, x_3) = ℯ^(-x_1*x_2) + 20x_3 + (10π - 3)/3

function F(x_1, x_2, x_3)

    f_1 = 3x_1 - cos(x_2*x_3) - 1/2
    f_2 = 4*x_1^2 - 625(x_2)^2 + 2*x_2 - 1
    f_3 = ℯ^(-x_1*x_2) + 20x_3 + (10π - 3)/3

    return [f_1; f_2; f_3]
end
#----------------------

F (generic function with 3 methods)

In [109]:
#Matriz Jacobiana:
function J(x_1, x_2, x_3)
    a_11 = 3
    a_12 = x_3*(sin(x_2*x_3))
    a_13 = x_2*(sin(x_2*x_3))

    a_21 = 8*x_1
    a_22 = -1250*(x_2) + 2
    a_23 = 0

    a_31 = -1*x_2*(ℯ^(-1*x_1*x_2))
    a_32 = -1*x_1*(ℯ^(-1*x_1*x_2))
    a_33 = 20

    return [a_11 a_12 a_13; a_21 a_22 a_23; a_31 a_32 a_33]
end

J (generic function with 2 methods)

In [110]:
function metodo_Broyden(x_1n, x_2n, x_3n)
    iteration = 0
    δ = 1
    x_n = [x_1n, x_2n, x_3n]

    #Primer iteración con Newton -> x_1
    x_m = metodo_Newton(x_1n, x_2n, x_3n)

    mJacobi = inv(J(x_1n, x_2n, x_3n)) #-> J(x_0)^-1

    iteration = 1

    while iteration ≤ max_iter && δ ≥ TOL

        mJacobi = sherman_morrison(x_n[1], x_n[2], x_n[3], x_m[1], x_m[2], x_m[3], mJacobi) #-> ≈ J(x_iteration)^-1
        x_n = [x_m[1], x_m[2], x_m[3]] #-> x_n = x_iteration
        
        #Se consigue a x_iteration+1
        x_m = x_n - (mJacobi * F(x_n[1], x_n[2], x_n[3]))

        #Error
        ω = abs.(x_m .- x_n)
        δ = max(ω[1], ω[2], ω[3])

        iteration = iteration + 1
    end

    println("La raíz encontrada es X_$iteration = $x_m")
    println("Evaluando la función original con la raíz encontrada: ")
    resultadoF = F(x_m[1], x_m[2], x_m[3])
    println("F(X_$iteration) = $resultadoF")
end

metodo_Broyden (generic function with 1 method)

In [111]:
#Valores iniciales
x_1n = 1
x_2n = 1
x_3n = 1

1

In [112]:
metodo_Broyden(x_1n, x_2n, x_3n)

La raíz encontrada es X_16 = [0.4999995325235557, 0.003199067572372998, -0.5235188629121877]
Evaluando la función original con la raíz encontrada: 
F(X_16) = [1.3301582058034e-12, -5.593037255557931e-9, 1.5045742429720121e-12]


**b.** $(2,1,-1)^t$

In [113]:
# ----- FUNCIONES -----
#Funciones originales
f_1(x_1, x_2, x_3) = x_1^2 + x_2 - 37
f_2(x_1, x_2, x_3) = x_1 - x_2^2 - 5
f_3(x_1, x_2, x_3) = x_1 + x_2 + x_3 - 3

function F(x_1, x_2, x_3)

    f_1 = x_1^2 + x_2 - 37
    f_2 = x_1 - x_2^2 - 5
    f_3 = x_1 + x_2 + x_3 - 3

    return [f_1; f_2; f_3]
end
#----------------------

F (generic function with 3 methods)

In [114]:
#Matriz Jacobiana:
function J(x_1, x_2, x_3)
    a_11 = 2*x_1
    a_12 = 1
    a_13 = 0

    a_21 = 1
    a_22 = -2*x_2
    a_23 = 0

    a_31 = 1
    a_32 = 1
    a_33 = 1

    return [a_11 a_12 a_13; a_21 a_22 a_23; a_31 a_32 a_33]
end

J (generic function with 2 methods)

In [115]:
function metodo_Broyden(x_1n, x_2n, x_3n)
    iteration = 0
    δ = 1
    x_n = [x_1n, x_2n, x_3n]

    #Primer iteración con Newton -> x_1
    x_m = metodo_Newton(x_1n, x_2n, x_3n)

    mJacobi = inv(J(x_1n, x_2n, x_3n)) #-> J(x_0)^-1

    iteration = 1

    while iteration ≤ max_iter && δ ≥ TOL

        mJacobi = sherman_morrison(x_n[1], x_n[2], x_n[3], x_m[1], x_m[2], x_m[3], mJacobi) #-> ≈ J(x_iteration)^-1
        x_n = [x_m[1], x_m[2], x_m[3]] #-> x_n = x_iteration
        
        #Se consigue a x_iteration+1
        x_m = x_n - (mJacobi * F(x_n[1], x_n[2], x_n[3]))

        #Error
        ω = abs.(x_m .- x_n)
        δ = max(ω[1], ω[2], ω[3])

        iteration = iteration + 1
    end

    println("La raíz encontrada es X_$iteration = $x_m")
    println("Evaluando la función original con la raíz encontrada: ")
    resultadoF = F(x_m[1], x_m[2], x_m[3])
    println("F(X_$iteration) = $resultadoF")
end

metodo_Broyden (generic function with 1 method)

In [116]:
#Valores iniciales
x_1n = 2
x_2n = 1
x_3n = -1

-1

In [117]:
metodo_Broyden(x_1n, x_2n, x_3n)

La raíz encontrada es X_8 = [5.999999930105521, 1.0000000146336137, -3.999999944739135]
Evaluando la función original con la raíz encontrada: 
F(X_8) = [-8.241001268061154e-7, -9.916170640167365e-8, 0.0]


**c.** $ (-1,-2,1)^t $

In [118]:
# ----- FUNCIONES -----
#Funciones originales
f_1(x_1, x_2, x_3) = x_1^3 + x_1^2*x_2 - x_1*x_3 + 6
f_2(x_1, x_2, x_3) = ℯ^(x_1) + ℯ^(x_2) - x_3 
f_3(x_1, x_2, x_3) = x_2^2 - 2*x_1*x_3 - 4

function F(x_1, x_2, x_3)

    f_1 = x_1^3 + x_1^2*x_2 - x_1*x_3 + 6
    f_2 = ℯ^(x_1) + ℯ^(x_2) - x_3 
    f_3 = x_2^2 - 2*x_1*x_3 - 4

    return [f_1; f_2; f_3]
end
#----------------------

F (generic function with 3 methods)

In [119]:
#Matriz Jacobiana:
function J(x_1, x_2, x_3)
    a_11 = 3*x_1^2 + 2*x_2*x_1 - x_3
    a_12 = x_1^2
    a_13 = -x_1

    a_21 = ℯ^(x_1)
    a_22 = ℯ^(x_2)
    a_23 = -1

    a_31 = -2*x_3
    a_32 = 2*x_2
    a_33 = -2*x_1

    return [a_11 a_12 a_13; a_21 a_22 a_23; a_31 a_32 a_33]
end

J (generic function with 2 methods)

In [120]:
function metodo_Broyden(x_1n, x_2n, x_3n)
    iteration = 0
    δ = 1
    x_n = [x_1n, x_2n, x_3n]

    #Primer iteración con Newton -> x_1
    x_m = metodo_Newton(x_1n, x_2n, x_3n)

    mJacobi = inv(J(x_1n, x_2n, x_3n)) #-> J(x_0)^-1

    iteration = 1

    while iteration ≤ max_iter && δ ≥ TOL

        mJacobi = sherman_morrison(x_n[1], x_n[2], x_n[3], x_m[1], x_m[2], x_m[3], mJacobi) #-> ≈ J(x_iteration)^-1
        x_n = [x_m[1], x_m[2], x_m[3]] #-> x_n = x_iteration
        
        #Se consigue a x_iteration+1
        x_m = x_n - (mJacobi * F(x_n[1], x_n[2], x_n[3]))

        #Error
        ω = abs.(x_m .- x_n)
        δ = max(ω[1], ω[2], ω[3])

        iteration = iteration + 1
    end

    println("La raíz encontrada es X_$iteration = $x_m")
    println("Evaluando la función original con la raíz encontrada: ")
    resultadoF = F(x_m[1], x_m[2], x_m[3])
    println("F(X_$iteration) = $resultadoF")
end

metodo_Broyden (generic function with 1 method)

In [121]:
#Valores iniciales
x_1n = -1
x_2n = -2
x_3n = 1

1

In [122]:
metodo_Broyden(x_1n, x_2n, x_3n)

La raíz encontrada es X_6 = [-1.456043002262457, -1.664230059731139, 0.4224934827403066]
Evaluando la función original con la raíz encontrada: 
F(X_6) = [-1.2493438790173172e-6, -4.9458842465988084e-8, -9.501962541236253e-7]


**d.** $ (0,0,0)^t $

In [123]:
# ----- FUNCIONES -----
#Funciones originales
f_1(x_1, x_2, x_3) = 6*x_1 - 2*cos(x_2*x_3) - 1
f_2(x_1, x_2, x_3) = 9*x_2 + √(x_1^2 + sin(x_3) + 1.06) + 0.9
f_3(x_1, x_2, x_3) = 60*x_3 + 3*ℯ^(-x_1*x_2) + 10π - 3

function F(x_1, x_2, x_3)

    f_1 = 6*x_1 - 2*cos(x_2*x_3) - 1
    f_2 = 9*x_2 + √(x_1^2 + sin(x_3) + 1.06) + 0.9
    f_3 = 60*x_3 + 3*ℯ^(-x_1*x_2) + 10π - 3

    return [f_1; f_2; f_3]
end
#----------------------

F (generic function with 3 methods)

In [124]:
#Matriz Jacobiana:
function J(x_1, x_2, x_3)
    a_11 = 6
    a_12 = 2*x_3*(sin(x_2*x_3))
    a_13 = 2*x_2*(sin(x_2*x_3))

    a_21 = x_1/(√(x_1^2 + sin(x_3) + 1.06))
    a_22 = 9
    a_23 = cos(x_3)/(2*√(x_1^2 + sin(x_3) + 1.06))

    a_31 = -3*x_2*ℯ^(-x_1*x_2)
    a_32 = -3*x_1*ℯ^(-x_1*x_2)
    a_33 = 60

    return [a_11 a_12 a_13; a_21 a_22 a_23; a_31 a_32 a_33]
end

J (generic function with 2 methods)

In [125]:
function metodo_Broyden(x_1n, x_2n, x_3n)
    iteration = 0
    δ = 1
    x_n = [x_1n, x_2n, x_3n]

    #Primer iteración con Newton -> x_1
    x_m = metodo_Newton(x_1n, x_2n, x_3n)

    mJacobi = inv(J(x_1n, x_2n, x_3n)) #-> J(x_0)^-1

    iteration = 1

    while iteration ≤ max_iter && δ ≥ TOL

        mJacobi = sherman_morrison(x_n[1], x_n[2], x_n[3], x_m[1], x_m[2], x_m[3], mJacobi) #-> ≈ J(x_iteration)^-1
        x_n = [x_m[1], x_m[2], x_m[3]] #-> x_n = x_iteration
        
        #Se consigue a x_iteration+1
        x_m = x_n - (mJacobi * F(x_n[1], x_n[2], x_n[3]))

        #Error
        ω = abs.(x_m .- x_n)
        δ = max(ω[1], ω[2], ω[3])

        iteration = iteration + 1
    end

    println("La raíz encontrada es X_$iteration = $x_m")
    println("Evaluando la función original con la raíz encontrada: ")
    resultadoF = F(x_m[1], x_m[2], x_m[3])
    println("F(X_$iteration) = $resultadoF")
end

metodo_Broyden (generic function with 1 method)

In [126]:
#Valores iniciales
x_1n = 0
x_2n = 0
x_3n = 0

0

In [127]:
metodo_Broyden(x_1n, x_2n, x_3n)

La raíz encontrada es X_3 = [0.49814467249652317, -0.19960588888315212, -0.5288259933186806]
Evaluando la función original con la raíz encontrada: 
F(X_3) = [-7.263776868882132e-8, 4.564340572166259e-8, -9.637106117565963e-7]


6.- El sistema no lineal

$$4x_1 - x_2 + x_3 = x_1x_4$$
$$-x_1 + 3x_2 - 2x_3 = x_2x_4$$
$$x_1 - 2x_2 + 3x_3 = x_3x_4$$
$$x_1^2 + x_2^2 + x_3^2 = 1$$

tiene seis soluciones.

**a.** Muestre que si $ (x_1, x_2, x_3, x_4)^t $ es una solución, entonces $ (-x_1, -x_2, -x_3, -x_4)^t $ es una solución.

**b.** Use el método de Broyden tres veces para aproximar cada solución. Itere hasta $||x^k-x^{k-1}||_\infty < 10^{-5}$.

In [128]:
# ----- FUNCIONES -----
#Funciones originales
f_1(x_1, x_2, x_3, x_4) = 4*x_1 - x_2 + x_3 - x_1*x_4
f_2(x_1, x_2, x_3, x_4) = -x_1 + 3*x_2 - 2*x_3 - x_2*x_4
f_3(x_1, x_2, x_3, x_4) = x_1 - 2*x_2 + 3*x_3 - x_3*x_4
f_3(x_1, x_2, x_3, x_4) = x_1^2 + x_2^2 + x_3^2 - 1

function F(x_1, x_2, x_3, x_4)

    f_1 = 4*x_1 - x_2 + x_3 - x_1*x_4
    f_2 = -x_1 + 3*x_2 - 2*x_3 - x_2*x_4
    f_3 = x_1 - 2*x_2 + 3*x_3 - x_3*x_4
    f_4 = x_1^2 + x_2^2 + x_3^2 - 1

    return [f_1; f_2; f_3; f_4]
end
#----------------------

F (generic function with 3 methods)

In [129]:
#Matriz Jacobiana:
function J(x_1, x_2, x_3, x_4)
    a_11 = 4 - x_4
    a_12 = -1
    a_13 = 1
    a_14 = x_1

    a_21 = -1
    a_22 = 3 - x_4
    a_23 = -2
    a_24 = x_2

    a_31 = 1
    a_32 = -2
    a_33 = 3 - x_4
    a_34 = x_3

    a_41 = 2*x_1
    a_42 = 2*x_2
    a_43 = 2*x_3
    a_44 = 0

    return [a_11 a_12 a_13 a_14; a_21 a_22 a_23 a_24; a_31 a_32 a_33 a_34; a_41 a_42 a_43 a_44]
end

J (generic function with 2 methods)

In [130]:
function metodo_Newton(x_1n, x_2n, x_3n, x_4n)
    vector_xm = 0.0
    iteration = 0

    while iteration ≤ 1
        vectorFx = F(x_1n, x_2n, x_3n, x_4n)  #Vector resultante al evaluar X_n en la función
        matrizJ = J(x_1n, x_2n, x_3n, x_4n)   #Matriz Jacobiana evaluada en el vector X_n

        #Se obtiene el resultado equivalente de multiplicar
        #la matriz Jacobiana invertida multiplicada por el 
        #vector resultante al evaluar F(X_n), recordando que 
        #esto es lo mismo que resolver un sistema de ecuaciones 
        #J(X_n)*Y = F(X_n), en donde se busca al vector Y:
        #vectorResul = metodoGaussJordan(matrizJ, vectorFx)
        #Sin embargo, obtenemos el mismo resultado haciendo esto:
        vectorResul = inv(matrizJ)*vectorFx

        vector_xn = [x_1n, x_2n, x_3n, x_4n]

        vector_xm = vector_xn .- vectorResul    #Se obtiene a X_1 = X_0 - Y

        #Se iguala X_0 a X_1 para la siguiente iteración
        x_1n = vector_xm[1]
        x_2n = vector_xm[2]
        x_3n = vector_xm[3]
        x_4n = vector_xm[4]

        iteration = iteration + 1
    end

    return vector_xm
end

metodo_Newton (generic function with 2 methods)

In [131]:
function sherman_morrison(x_1n, x_2n, x_3n, x_4n, x_1m, x_2m, x_3m, x_4m, A_0)
    x0 = [x_1n, x_2n, x_3n, x_4n]
    x1 = [x_1m, x_2m, x_3m, x_4m]
    y_i = F(x_1m, x_2m, x_3m, x_4m) .- F(x_1n, x_2n, x_3n, x_4n)
    s_i = x1 .- x0

    #Producto externo:
    AxY = A_0 * y_i
    prodExtern = 0.0
    for j in 1:convert(Int64, size(s_i)[1])
        prodExtern = prodExtern + s_i[j] * AxY[j]
    end

    A_1 = A_0 + ((s_i - (A_0 * y_i)) * (transpose(s_i) * A_0)) / prodExtern
    return A_1
end

sherman_morrison (generic function with 2 methods)

In [132]:
function metodo_Broyden(x_1n, x_2n, x_3n, x_4n)
    iteration = 0
    δ = 1
    x_n = [x_1n, x_2n, x_3n, x_4n]

    #Primer iteración con Newton -> x_1
    x_m = metodo_Newton(x_1n, x_2n, x_3n, x_4n)

    mJacobi = inv(J(x_1n, x_2n, x_3n, x_4n)) #-> J(x_0)^-1

    iteration = 1

    while iteration ≤ max_iter && δ ≥ TOL

        mJacobi = sherman_morrison(x_n[1], x_n[2], x_n[3], x_n[4], x_m[1], x_m[2], x_m[3], x_m[4], mJacobi) #-> ≈ J(x_iteration)^-1
        x_n = [x_m[1], x_m[2], x_m[3], x_m[4]] #-> x_n = x_iteration
        
        #Se consigue a x_iteration+1
        x_m = x_n - (mJacobi * F(x_n[1], x_n[2], x_n[3], x_n[4]))

        #Error
        ω = abs.(x_m .- x_n)
        δ = max(ω[1], ω[2], ω[3], ω[4])

        iteration = iteration + 1
    end

    println("La raíz encontrada es X_$iteration = $x_m")
    println("Evaluando la función original con la raíz encontrada: ")
    resultadoF = F(x_m[1], x_m[2], x_m[3], x_m[4])
    println("F(X_$iteration) = $resultadoF")
end

metodo_Broyden (generic function with 2 methods)

**Resolución:**

In [133]:
#Valores que acercan las dos primeras soluciones{
    x_1n = 1.0
    x_2n = 1.0
    x_3n = 1.0
    x_4n = 1.0 

#=     x_1n = -1.0
    x_2n = -1.0
    x_3n = -1.0
    x_4n = 1.0 =#
#}

1.0

In [134]:
metodo_Broyden(x_1n, x_2n, x_3n, x_4n)

La raíz encontrada es X_5 = [-3.6787267876184515e-16, 0.7071067813444347, 0.7071067813444349, 1.0000000000000024]
Evaluando la función original con la raíz encontrada: 
F(X_5) = [-8.533726483258262e-16, -1.9984014443252818e-15, -1.7763568394002505e-15, 4.4657255671154417e-10]


In [135]:
#Valores que acercan las dos primeras soluciones{
#=     x_1n = 1.0
    x_2n = 1.0
    x_3n = 1.0
    x_4n = 1.0  =#

    x_1n = -1.0
    x_2n = -1.0
    x_3n = -1.0
    x_4n = 1.0
#}

1.0

In [136]:
metodo_Broyden(x_1n, x_2n, x_3n, x_4n)

La raíz encontrada es X_5 = [3.6787267876184515e-16, -0.7071067813444347, -0.7071067813444349, 1.0000000000000024]
Evaluando la función original con la raíz encontrada: 
F(X_5) = [8.533726483258262e-16, 1.9984014443252818e-15, 1.7763568394002505e-15, 4.4657255671154417e-10]


In [137]:
#Valores que acercan las dos segundas soluciones{
    x_1n = -1
    x_2n = -0.5
    x_3n = 0.5
    x_4n = 3.5

#=     x_1n = 1
    x_2n = 0.5
    x_3n = -0.5
    x_4n = 3.5 =#
#}

3.5

In [138]:
metodo_Broyden(x_1n, x_2n, x_3n, x_4n)

La raíz encontrada es X_7 = [-0.8164965889835657, -0.40824829449178296, 0.40824829449178296, 3.000000004481897]
Evaluando la función original con la raíz encontrada: 
F(X_7) = [3.6594536290124324e-9, 1.8297265924616113e-9, -1.8297265924616113e-9, 1.97326968098821e-8]


In [139]:
#Valores que acercan las dos segundas soluciones{
#=     x_1n = -1
    x_2n = -0.5
    x_3n = 0.5
    x_4n = 3.5 =#

    x_1n = 1
    x_2n = 0.5
    x_3n = -0.5
    x_4n = 3.5
#}

3.5

In [140]:
metodo_Broyden(x_1n, x_2n, x_3n, x_4n)

La raíz encontrada es X_7 = [0.8164965889835657, 0.40824829449178296, -0.40824829449178296, 3.000000004481897]
Evaluando la función original con la raíz encontrada: 
F(X_7) = [-3.6594536290124324e-9, -1.8297265924616113e-9, 1.8297265924616113e-9, 1.97326968098821e-8]


In [141]:
#Valores que acercan las dos segundas soluciones{
    x_1n = 1
    x_2n = -1
    x_3n = 1
    x_4n = 7

#=     x_1n = -1
    x_2n = 1
    x_3n = -1
    x_4n = 7 =#
#}

7

In [142]:
metodo_Broyden(x_1n, x_2n, x_3n, x_4n)

La raíz encontrada es X_9 = [0.5773503127143458, -0.5773503127143459, 0.5773503127143459, 6.000000002148058]
Evaluando la función original con la raíz encontrada: 
F(X_9) = [-1.2401812909956789e-9, 1.2401821791740986e-9, -1.2401821791740986e-9, 1.5077405879360128e-7]


In [143]:
#Valores que acercan las dos segundas soluciones{
#=     x_1n = 1
    x_2n = -1
    x_3n = 1
    x_4n = 7 =#

    x_1n = -1
    x_2n = 1
    x_3n = -1
    x_4n = 7
#}

7

In [144]:
metodo_Broyden(x_1n, x_2n, x_3n, x_4n)

La raíz encontrada es X_9 = [-0.5773503127143458, 0.5773503127143459, -0.5773503127143459, 6.000000002148058]
Evaluando la función original con la raíz encontrada: 
F(X_9) = [1.2401812909956789e-9, -1.2401821791740986e-9, 1.2401821791740986e-9, 1.5077405879360128e-7]


De esa forma, tenemos que las seis soluciones del sistema:

$$ S_1 = (-0.0000, 0.7071, 0.7071, 1.0000)^t $$
$$ S_2 = (0.0000, -0.7071, -0.7071, 1.0000)^t $$
$$ S_3 = (-0.8164, -0.4082, 0.4082, 3.0000)^t $$
$$ S_4 = (0.8164, 0.4082, -0.4082, 3.0000)^t $$
$$ S_5 = (0.5773, -0.5773, 0.5773, 6.0000)^t $$
$$ S_6 = (-0.5773, 0.5773, -0.5773, 6.0000)^t $$

Así, resolvimos tanto el inciso **a** como el **b**, pues demostramos que si $ (x_1, x_2, x_3, x_4)^t $ es una solución, entonces, efectivamente $ (-x_1, -x_2, -x_3, -x_4)^t $ es una solución, y además, aproximamos cada solución con un error $<10^{-5}$.